// yt-cast-receiver/dist/lib/utils/AsyncTaskQueue.js
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _AsyncTaskQueue_tasks, _AsyncTaskQueue_stopped, _AsyncTaskQueue_autostart;
/**
 * @internal
 *
 * A queue that takes {@link Task} objects and executes them in the same
 * order as they are added.
 */
class AsyncTaskQueue {
    constructor() {
        _AsyncTaskQueue_tasks.set(this, void 0);
        _AsyncTaskQueue_stopped.set(this, void 0);
        _AsyncTaskQueue_autostart.set(this, void 0);
        __classPrivateFieldSet(this, _AsyncTaskQueue_stopped, true, "f");
        __classPrivateFieldSet(this, _AsyncTaskQueue_tasks, [], "f");
        __classPrivateFieldSet(this, _AsyncTaskQueue_autostart, true, "f");
    }
    setAutoStart(value) {
        __classPrivateFieldSet(this, _AsyncTaskQueue_autostart, value, "f");
        if (value && __classPrivateFieldGet(this, _AsyncTaskQueue_tasks, "f").length > 0) {
            void this.start();
        }
    }
    push(task) {
        __classPrivateFieldGet(this, _AsyncTaskQueue_tasks, "f").push(task);
        if (__classPrivateFieldGet(this, _AsyncTaskQueue_stopped, "f") && __classPrivateFieldGet(this, _AsyncTaskQueue_autostart, "f")) {
            void this.start();
        }
    }
    unshift(task) {
        __classPrivateFieldGet(this, _AsyncTaskQueue_tasks, "f").unshift(task);
    }
    async start() {
        __classPrivateFieldSet(this, _AsyncTaskQueue_stopped, false, "f");
        let failedTask = null;
        while (__classPrivateFieldGet(this, _AsyncTaskQueue_tasks, "f").length > 0 && !__classPrivateFieldGet(this, _AsyncTaskQueue_stopped, "f")) {
            const task = __classPrivateFieldGet(this, _AsyncTaskQueue_tasks, "f").shift();
            try {
                if (task)
                    await task.run();
            }
            catch (error) {
                failedTask = { task, error };
                break;
            }
        }
        __classPrivateFieldSet(this, _AsyncTaskQueue_stopped, true, "f");
        if (failedTask && failedTask.task.onError) {
            failedTask.task.onError(failedTask.task, failedTask.error);
        }
    }
    clear() {
        if (!__classPrivateFieldGet(this, _AsyncTaskQueue_stopped, "f")) {
            __classPrivateFieldSet(this, _AsyncTaskQueue_stopped, true, "f");
        }
        __classPrivateFieldGet(this, _AsyncTaskQueue_tasks, "f").forEach((task) => task.cancel());
        __classPrivateFieldSet(this, _AsyncTaskQueue_tasks, [], "f");
    }
    stop() {
        __classPrivateFieldSet(this, _AsyncTaskQueue_stopped, true, "f");
    }
    forEach(callbackfn, thisArg) {
        return __classPrivateFieldGet(this, _AsyncTaskQueue_tasks, "f").forEach(callbackfn, thisArg);
    }
}
_AsyncTaskQueue_tasks = new WeakMap(), _AsyncTaskQueue_stopped = new WeakMap(), _AsyncTaskQueue_autostart = new WeakMap();
export default AsyncTaskQueue;
