// yt-cast-receiver/dist/lib/utils/Errors.js
export default class YouTubeCastReceiverError extends Error {
    constructor(message, cause, info) {
        super(message);
        this.name = 'YouTubeCastReceiverError';
        if (cause) {
            this.cause = cause;
        }
        if (info) {
            this.info = info;
        }
    }
    getCauses() {
        const causes = [];
        let c = this.cause;
        while (c) {
            causes.push(c);
            c = c instanceof YouTubeCastReceiverError ? c.cause : null;
        }
        return causes;
    }
}
export class ConnectionError extends YouTubeCastReceiverError {
    constructor(message, url, cause) {
        super(message, cause, { url });
        this.name = 'ConnectionError';
    }
}
export class AbortError extends YouTubeCastReceiverError {
    constructor(message, url) {
        super(message, undefined, { url });
        this.name = 'AbortError';
    }
}
export class BadResponseError extends YouTubeCastReceiverError {
    constructor(message, url, response) {
        super(message, undefined, { url, response });
        this.name = 'BadResponseError';
    }
}
export class DataError extends YouTubeCastReceiverError {
    constructor(message, cause, data) {
        super(message, cause, { data });
        this.name = 'DataError';
    }
}
export class IncompleteAPIDataError extends YouTubeCastReceiverError {
    constructor(message, missing) {
        super(message, undefined, { missing });
        this.name = 'IncompleteAPIDataError';
    }
}
export class SessionError extends YouTubeCastReceiverError {
    constructor(message, cause) {
        super(message, cause);
        this.name = 'SessionError';
    }
}
export class AppError extends YouTubeCastReceiverError {
    constructor(message, cause) {
        super(message, cause);
        this.name = 'AppError';
    }
}
export class DialServerError extends YouTubeCastReceiverError {
    constructor(message, cause) {
        super(message, cause);
        this.name = 'DialServerError';
    }
}
export class SenderConnectionError extends YouTubeCastReceiverError {
    constructor(message, cause, action) {
        super(message, cause, { action: action || 'unknown' });
        this.name = 'SenderConnectionError';
    }
}
