// yt-cast-receiver/dist/lib/Player.js
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _Player_instances, _Player_status, _Player_queue, _Player_cpn, _Player_logger, _Player_previousState, _Player_zeroVolumeLevelOnMute, _Player_setStatusAndEmit;
import EventEmitter from 'events';
import { v4 as uuidv4 } from 'uuid';
import { AUTOPLAY_MODES, PLAYER_STATUSES } from './Constants.js';
import Playlist from './app/Playlist.js';
/**
 * `Player` abstract class that leaves playback functionality to implementors.
 */
class Player extends EventEmitter {
    constructor() {
        super();
        _Player_instances.add(this);
        _Player_status.set(this, void 0);
        _Player_queue.set(this, void 0);
        _Player_cpn.set(this, void 0);
        _Player_logger.set(this, void 0);
        _Player_previousState.set(this, void 0);
        _Player_zeroVolumeLevelOnMute.set(this, void 0);
        __classPrivateFieldSet(this, _Player_status, PLAYER_STATUSES.IDLE, "f");
        __classPrivateFieldSet(this, _Player_queue, new Playlist(), "f");
        __classPrivateFieldSet(this, _Player_cpn, uuidv4().replace(/-/g, '').substring(0, 16), "f");
        __classPrivateFieldSet(this, _Player_previousState, null, "f");
        __classPrivateFieldSet(this, _Player_zeroVolumeLevelOnMute, false, "f");
    }
    /**
     * @internal
     */
    setLogger(logger) {
        __classPrivateFieldSet(this, _Player_logger, logger, "f");
    }
    /**
     * Notifies senders that player is in 'loading' state, then calls `doPlay()`;
     * if returned Promise resolves to `true`, notifies senders that playback has started.
     * @param video - The target video to play.
     * @param position - The position (in seconds) from which to start playback.
     * @param AID - Internal use; do not specify.
     * @returns Promise that resolves to the resolved result of `doPlay()`.
     */
    async play(video, position, AID) {
        if (this.status === PLAYER_STATUSES.PLAYING) {
            await this.stop();
        }
        __classPrivateFieldGet(this, _Player_logger, "f").info(`[yt-cast-receiver] Player.play(): ${video.id} @ ${position || 0}s`);
        __classPrivateFieldGet(this, _Player_queue, "f").setAsCurrent(video);
        await __classPrivateFieldGet(this, _Player_instances, "m", _Player_setStatusAndEmit).call(this, PLAYER_STATUSES.LOADING, AID);
        const result = await this.doPlay(video, position || 0);
        if (result) {
            await __classPrivateFieldGet(this, _Player_instances, "m", _Player_setStatusAndEmit).call(this, PLAYER_STATUSES.PLAYING, AID);
        }
        else {
            // Notify sender apps of stopped state so they don't display the loading spinner
            await __classPrivateFieldGet(this, _Player_instances, "m", _Player_setStatusAndEmit).call(this, PLAYER_STATUSES.STOPPED, AID);
        }
        return result;
    }
    /**
     * Calls `doPause()`; if returned Promise resolves to `true`, notifies connected senders that playback has paused.
     * @param AID - Internal use; do not specify.
     * @returns Promise that resolves to the resolved result of `doPause()`, or `false` if no playback is in progress.
     */
    async pause(AID) {
        __classPrivateFieldGet(this, _Player_logger, "f").info('[yt-cast-receiver] Player.pause()');
        if (this.status === PLAYER_STATUSES.PLAYING) {
            const result = await this.doPause();
            if (result) {
                await __classPrivateFieldGet(this, _Player_instances, "m", _Player_setStatusAndEmit).call(this, PLAYER_STATUSES.PAUSED, AID);
            }
            return result;
        }
        return false;
    }
    /**
     * Calls `doResume()`; if returned Promise resolves to `true`, notifies connected senders that playback has resumed.
     * @param AID - Internal use; do not specify.
     * @returns Promise that resolves to the resolved result of `doResume()`, or `false` if player is not in paused state.
     */
    async resume(AID) {
        __classPrivateFieldGet(this, _Player_logger, "f").info('[yt-cast-receiver] Player.resume()');
        if (this.status === PLAYER_STATUSES.PLAYING) {
            await __classPrivateFieldGet(this, _Player_instances, "m", _Player_setStatusAndEmit).call(this, PLAYER_STATUSES.PLAYING, AID);
            return false;
        }
        else if (this.status === PLAYER_STATUSES.PAUSED) {
            const result = await this.doResume();
            if (result) {
                await __classPrivateFieldGet(this, _Player_instances, "m", _Player_setStatusAndEmit).call(this, PLAYER_STATUSES.PLAYING, AID);
            }
            return result;
        }
        else if (__classPrivateFieldGet(this, _Player_queue, "f").current) {
            return this.play(__classPrivateFieldGet(this, _Player_queue, "f").current);
        }
        return false;
    }
    /**
     * Calls `doStop()`; if returned Promise resolves to `true`, notifies connected senders that playback has stopped.
     * @param AID - Internal use; do not specify.
     * @returns A Promise that resolves to the result of `doStop()`; `true` if player already in stopped or idle state.
     */
    async stop(AID) {
        if (this.status === PLAYER_STATUSES.STOPPED || this.status === PLAYER_STATUSES.IDLE) {
            return true;
        }
        __classPrivateFieldGet(this, _Player_logger, "f").info('[yt-cast-receiver] Player.stop()');
        const result = await this.doStop();
        if (result) {
            await __classPrivateFieldGet(this, _Player_instances, "m", _Player_setStatusAndEmit).call(this, PLAYER_STATUSES.STOPPED, AID);
        }
        return result;
    }
    /**
     * Calls `doSeek()`; if returned Promise resolves to `true`, notifies connected senders of new seek position.
     * @param position - The position, in seconds, to seek to.
     * @param AID - Internal use; do not specify.
     * @returns Promise that resolves to the resolved result of `doSeek()`; `false` if no playback is in progress or otherwise not in paused state.
     */
    async seek(position, AID) {
        if (__classPrivateFieldGet(this, _Player_status, "f") !== PLAYER_STATUSES.PLAYING && __classPrivateFieldGet(this, _Player_status, "f") !== PLAYER_STATUSES.PAUSED) {
            return false;
        }
        __classPrivateFieldGet(this, _Player_logger, "f").info(`[yt-cast-receiver] Player.seek(): ${position}s`);
        const previousState = await this.getState();
        const fakeState = { ...previousState, position, status: PLAYER_STATUSES.LOADING };
        this.emit('state', { current: fakeState, previous: previousState, AID });
        const result = await this.doSeek(position);
        if (result) {
            __classPrivateFieldSet(this, _Player_previousState, fakeState, "f");
            if (previousState.status === PLAYER_STATUSES.PAUSED) {
                __classPrivateFieldSet(this, _Player_status, PLAYER_STATUSES.PAUSED, "f");
                return this.resume(AID);
            }
            await __classPrivateFieldGet(this, _Player_instances, "m", _Player_setStatusAndEmit).call(this, PLAYER_STATUSES.PLAYING, AID);
        }
        return result;
    }
    /**
     * Plays the next video in the player queue. If already reached end of queue,
     * play autoplay video if available. Notifies senders on successful playback.
     * @param AID - Internal use; do not specify.
     * @returns Promise that resolves to `true` on playback of the next video; `false` otherwise.
     */
    async next(AID) {
        if (__classPrivateFieldGet(this, _Player_queue, "f").isUpdating) {
            __classPrivateFieldGet(this, _Player_logger, "f").debug('[yt-cast-receiver] Player.next() ignored: queue is updating.');
            return false;
        }
        __classPrivateFieldGet(this, _Player_logger, "f").info('[yt-cast-receiver] Player.next()');
        const nextVideo = await __classPrivateFieldGet(this, _Player_queue, "f").next();
        if (!nextVideo) {
            __classPrivateFieldGet(this, _Player_logger, "f").info('[yt-cast-receiver] No next video in queue.');
            if (this.autoplayMode === AUTOPLAY_MODES.ENABLED) {
                const autoplayVideo = __classPrivateFieldGet(this, _Player_queue, "f").autoplay;
                if (autoplayVideo) {
                    __classPrivateFieldGet(this, _Player_logger, "f").info(`[yt-cast-receiver] Play autoplay video: ${autoplayVideo.id}.`);
                    return this.play(autoplayVideo, 0, AID);
                }
                __classPrivateFieldGet(this, _Player_logger, "f").info('[yt-cast-receiver] No autoplay video available.');
            }
            else {
                __classPrivateFieldGet(this, _Player_logger, "f").info('[yt-cast-receiver] No autoplay video - autoplay is disabled.');
            }
            await this.stop(AID);
            return false;
        }
        return this.play(nextVideo);
    }
    /**
     * Plays the previous video in the player queue. Notifies senders on successful playback.
     * @param AID - Internal use; do not specify.
     * @returns Promise that resolves to `true` on playback of the previous video; `false` otherwise.
     */
    async previous(AID) {
        if (__classPrivateFieldGet(this, _Player_queue, "f").isUpdating) {
            __classPrivateFieldGet(this, _Player_logger, "f").debug('[yt-cast-receiver] Player.previous() ignored: queue is updating.');
            return false;
        }
        __classPrivateFieldGet(this, _Player_logger, "f").info('[yt-cast-receiver] Player.previous()');
        const previousVideo = await __classPrivateFieldGet(this, _Player_queue, "f").previous();
        if (!previousVideo) {
            __classPrivateFieldGet(this, _Player_logger, "f").info('[yt-cast-receiver] No previous video in queue.');
            await this.stop(AID);
            return false;
        }
        return await this.play(previousVideo, 0, AID);
    }
    /**
     * Calls `doSetVolume()`; if returned Promise resolves to `true`, notifies connected senders of new volume level.
     * @param volume - (object)
     *   - `level`: (number) volume level between 0-100.
     *   - `muted`: (boolean) muted state.
     * @param AID - Internal use; do not specify.
     * @returns Promise that resolves to the resolved result of `doSetVolume()`.
     */
    async setVolume(volume, AID) {
        const v = { ...volume };
        if (__classPrivateFieldGet(this, _Player_zeroVolumeLevelOnMute, "f") && volume.muted && volume.level > 0) {
            __classPrivateFieldGet(this, _Player_logger, "f").debug('[yt-cast-receiver] Enforcing \'zeroVolumeLevelOnMute: true\'...');
            v.level = 0;
        }
        __classPrivateFieldGet(this, _Player_logger, "f").info('[yt-cast-receiver] Player.setVolume():', v);
        const previousState = await this.getState();
        const result = await this.doSetVolume(v);
        if (result) {
            const currentState = await this.getState();
            this.emit('state', { current: currentState, previous: previousState, AID });
        }
        return result;
    }
    /**
     * Resets the player to Idle state.
     * @param AID - Internal use; do not specify.
     */
    async reset(AID) {
        __classPrivateFieldGet(this, _Player_logger, "f").info('[yt-cast-receiver] Player.reset()');
        __classPrivateFieldGet(this, _Player_queue, "f").reset();
        await this.stop(AID);
        __classPrivateFieldSet(this, _Player_previousState, null, "f");
        __classPrivateFieldGet(this, _Player_instances, "m", _Player_setStatusAndEmit).call(this, PLAYER_STATUSES.IDLE, AID)
            .catch((error) => this.logger.error('[yt-cast-receiver] Caught error emitting status after player reset:', error));
    }
    /**
     * Calls `doGetVolume()`.
     * @returns Promise that resolves to the resolved result of `doGetVolume()`.
     */
    async getVolume() {
        const v = await this.doGetVolume();
        return {
            level: v.level < 0 ? 0 : v.level > 100 ? 100 : v.level,
            muted: v.muted
        };
    }
    /**
     * Calls `doGetPosition()`.
     * @returns Promise returned by `doGetPosition()`.
     */
    getPosition() {
        return this.doGetPosition();
    }
    /**
     * Calls `doGetDuration()`
     * @returns Promise returned by `doGetDuration()`.
     */
    getDuration() {
        return this.doGetDuration();
    }
    /**
     * @internal
     */
    async setZeroVolumeLevelOnMute(value, AID) {
        __classPrivateFieldSet(this, _Player_zeroVolumeLevelOnMute, value, "f");
        const currentVolume = await this.getVolume();
        if (__classPrivateFieldGet(this, _Player_zeroVolumeLevelOnMute, "f") && currentVolume.muted && currentVolume.level > 0) {
            __classPrivateFieldGet(this, _Player_logger, "f").debug(`[yt-cast-receiver] Calling setVolume() to enforce 'zeroVolumeLevelOnMute: ${value}'...`);
            await this.setVolume(currentVolume, AID);
        }
    }
    get logger() {
        return __classPrivateFieldGet(this, _Player_logger, "f");
    }
    get status() {
        return __classPrivateFieldGet(this, _Player_status, "f");
    }
    get autoplayMode() {
        return __classPrivateFieldGet(this, _Player_queue, "f").autoplayMode;
    }
    get cpn() {
        return __classPrivateFieldGet(this, _Player_cpn, "f");
    }
    get queue() {
        return __classPrivateFieldGet(this, _Player_queue, "f");
    }
    get zeroVolumeLevelOnMute() {
        return __classPrivateFieldGet(this, _Player_zeroVolumeLevelOnMute, "f");
    }
    getNavInfo() {
        return {
            hasPrevious: __classPrivateFieldGet(this, _Player_queue, "f").hasPrevious,
            hasNext: __classPrivateFieldGet(this, _Player_queue, "f").hasNext,
            autoplayMode: this.autoplayMode
        };
    }
    async getState() {
        return {
            status: this.status,
            queue: __classPrivateFieldGet(this, _Player_queue, "f").getState(),
            position: await this.getPosition(),
            duration: await this.getDuration(),
            volume: await this.getVolume(),
            cpn: __classPrivateFieldGet(this, _Player_cpn, "f")
        };
    }
    /**
     * Signals that there has been a change in player state that is not captured elsewhere
     * in the `Player` implementation. This method will update the `Player` instance's
     * internal state and, if necessary, notifies senders of the new player state.
     * @param newStatus - The new player status; `undefined` for no change in player status.
     */
    async notifyExternalStateChange(newStatus) {
        return __classPrivateFieldGet(this, _Player_instances, "m", _Player_setStatusAndEmit).call(this, newStatus);
    }
    on(event, listener) {
        super.on(event, listener);
        return this;
    }
}
_Player_status = new WeakMap(), _Player_queue = new WeakMap(), _Player_cpn = new WeakMap(), _Player_logger = new WeakMap(), _Player_previousState = new WeakMap(), _Player_zeroVolumeLevelOnMute = new WeakMap(), _Player_instances = new WeakSet(), _Player_setStatusAndEmit = 
/**
 * Updates the player status to `status` and emits a `state` event to indicate a change in player state.
 * @param status - Current player status; `undefined` for no change in player status.
 * @param AID
 */
async function _Player_setStatusAndEmit(status, AID) {
    const previous = __classPrivateFieldGet(this, _Player_previousState, "f");
    if (status) {
        __classPrivateFieldSet(this, _Player_status, status, "f");
    }
    const current = __classPrivateFieldSet(this, _Player_previousState, await this.getState(), "f");
    this.emit('state', { AID, current, previous });
};
export default Player;
