// yt-cast-receiver/dist/lib/YouTubeCastReceiver.js
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _YouTubeCastReceiver_app, _YouTubeCastReceiver_server, _YouTubeCastReceiver_logger, _YouTubeCastReceiver_status;
import EventEmitter from 'events';
import YouTubeApp from './app/YouTubeApp.js';
import DialServer from './dial/DialServer.js';
import DefaultLogger from './utils/DefaultLogger.js';
import { STATUSES } from './Constants.js';
import { hostname } from 'os';
import DefaultDataStore from './utils/DefaultDataStore.js';
/**
 * Main class of `yt-cast-receiver` library.
 *
 * To create a `YouTubeCastReceiver` instance, you need to provide at least a
 * {@link Player} implementation.
 */
class YouTubeCastReceiver extends EventEmitter {
    constructor(player, options = {}) {
        super();
        _YouTubeCastReceiver_app.set(this, void 0);
        _YouTubeCastReceiver_server.set(this, void 0);
        _YouTubeCastReceiver_logger.set(this, void 0);
        _YouTubeCastReceiver_status.set(this, void 0);
        __classPrivateFieldSet(this, _YouTubeCastReceiver_logger, options.logger || new DefaultLogger(), "f");
        __classPrivateFieldGet(this, _YouTubeCastReceiver_logger, "f").setLevel(options.logLevel || 'info');
        const friendlyName = options.device?.name || hostname();
        const screenName = options.device?.screenName || `YouTube on ${friendlyName}`;
        const dataStore = options.dataStore !== false ? (options.dataStore || new DefaultDataStore()) : null;
        if (dataStore) {
            dataStore.setLogger(__classPrivateFieldGet(this, _YouTubeCastReceiver_logger, "f"));
        }
        const appOptions = {
            ...(options.app || {}),
            screenName,
            brand: options.device?.brand,
            model: options.device?.model,
            dataStore,
            logger: __classPrivateFieldGet(this, _YouTubeCastReceiver_logger, "f")
        };
        const dialOptions = {
            ...(options.dial || {}),
            friendlyName,
            manufacturer: options.device?.brand,
            modelName: options.device?.model,
            logger: __classPrivateFieldGet(this, _YouTubeCastReceiver_logger, "f")
        };
        __classPrivateFieldSet(this, _YouTubeCastReceiver_app, new YouTubeApp(player, appOptions), "f");
        __classPrivateFieldSet(this, _YouTubeCastReceiver_server, new DialServer(__classPrivateFieldGet(this, _YouTubeCastReceiver_app, "f"), dialOptions), "f");
        __classPrivateFieldGet(this, _YouTubeCastReceiver_app, "f").on('senderConnect', (sender) => {
            this.emit('senderConnect', sender);
        });
        __classPrivateFieldGet(this, _YouTubeCastReceiver_app, "f").on('senderDisconnect', (sender, implicit) => {
            this.emit('senderDisconnect', sender, implicit);
        });
        __classPrivateFieldGet(this, _YouTubeCastReceiver_app, "f").on('error', (error) => {
            this.emit('error', error);
        });
        __classPrivateFieldGet(this, _YouTubeCastReceiver_app, "f").on('terminate', (error) => {
            void (async () => {
                try {
                    await this.stop();
                }
                catch (_error) {
                    // Do nothing - we're quitting anyway
                }
                __classPrivateFieldGet(this, _YouTubeCastReceiver_logger, "f").error('[yt-cast-receiver] Receiver terminated due to error:', error);
                this.emit('terminate', error);
            })();
        });
        __classPrivateFieldSet(this, _YouTubeCastReceiver_status, STATUSES.STOPPED, "f");
    }
    async start() {
        if (this.status !== STATUSES.STOPPED) {
            __classPrivateFieldGet(this, _YouTubeCastReceiver_logger, "f").warn('[yt-cast-receiver] start() called but receiver not in STOPPED state.');
            return;
        }
        __classPrivateFieldSet(this, _YouTubeCastReceiver_status, STATUSES.STARTING, "f");
        try {
            await __classPrivateFieldGet(this, _YouTubeCastReceiver_app, "f").start();
            await __classPrivateFieldGet(this, _YouTubeCastReceiver_server, "f").start();
        }
        catch (error) {
            __classPrivateFieldGet(this, _YouTubeCastReceiver_logger, "f").error('[yt-cast-receiver] Failed to start receiver:', error);
            try {
                if (__classPrivateFieldGet(this, _YouTubeCastReceiver_app, "f").state === STATUSES.RUNNING) {
                    await __classPrivateFieldGet(this, _YouTubeCastReceiver_app, "f").stop();
                }
                if (__classPrivateFieldGet(this, _YouTubeCastReceiver_server, "f").status === STATUSES.RUNNING) {
                    await __classPrivateFieldGet(this, _YouTubeCastReceiver_server, "f").stop();
                }
            }
            catch (_error) {
                // Do nothing
            }
            __classPrivateFieldSet(this, _YouTubeCastReceiver_status, STATUSES.STOPPED, "f");
            throw error;
        }
        __classPrivateFieldSet(this, _YouTubeCastReceiver_status, STATUSES.RUNNING, "f");
    }
    async stop() {
        if (this.status !== STATUSES.RUNNING) {
            __classPrivateFieldGet(this, _YouTubeCastReceiver_logger, "f").warn('[yt-cast-receiver] stop() called but receiver not in RUNNING state.');
            return;
        }
        __classPrivateFieldSet(this, _YouTubeCastReceiver_status, STATUSES.STOPPING, "f");
        try {
            await __classPrivateFieldGet(this, _YouTubeCastReceiver_app, "f").stop();
            await __classPrivateFieldGet(this, _YouTubeCastReceiver_server, "f").stop();
        }
        catch (error) {
            __classPrivateFieldSet(this, _YouTubeCastReceiver_status, STATUSES.RUNNING, "f");
            __classPrivateFieldGet(this, _YouTubeCastReceiver_logger, "f").error('[yt-cast-receiver] Failed to stop receiver:', error);
            throw error;
        }
        __classPrivateFieldSet(this, _YouTubeCastReceiver_status, STATUSES.STOPPED, "f");
    }
    enableAutoplayOnConnect(value) {
        __classPrivateFieldGet(this, _YouTubeCastReceiver_app, "f").enableAutoplayOnConnect(value);
    }
    setResetPlayerOnDisconnectPolicy(value) {
        __classPrivateFieldGet(this, _YouTubeCastReceiver_app, "f").setResetPlayerOnDisconnectPolicy(value);
    }
    setLogLevel(value) {
        __classPrivateFieldGet(this, _YouTubeCastReceiver_logger, "f").setLevel(value);
    }
    getPairingCodeRequestService() {
        return __classPrivateFieldGet(this, _YouTubeCastReceiver_app, "f").getPairingCodeRequestService();
    }
    getConnectedSenders() {
        return __classPrivateFieldGet(this, _YouTubeCastReceiver_app, "f").connectedSenders;
    }
    get status() {
        return __classPrivateFieldGet(this, _YouTubeCastReceiver_status, "f");
    }
    get logger() {
        return __classPrivateFieldGet(this, _YouTubeCastReceiver_logger, "f");
    }
    emit(event, ...args) {
        return super.emit(event, ...args);
    }
    on(event, listener) {
        super.on(event, listener);
        return this;
    }
}
_YouTubeCastReceiver_app = new WeakMap(), _YouTubeCastReceiver_server = new WeakMap(), _YouTubeCastReceiver_logger = new WeakMap(), _YouTubeCastReceiver_status = new WeakMap();
export default YouTubeCastReceiver;
