// yt-cast-receiver/dist/lib/app/BindParams.js
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _BindParams_instances, _BindParams_generateZX, _BindParams_generateRID;
import { v4 as uuidv4 } from 'uuid';
import queryString from 'query-string';
import { IncompleteAPIDataError } from '../utils/Errors.js';
/** @internal */
const DEFAULT_DEVICE_INFO = {
    year: 0,
    os: 'Windows',
    osVersion: '10.0',
    chipset: '',
    clientName: 'TVHTML5',
    dialAdditionalDataSupportLevel: 'unsupported',
    mdxDialServerType: 'MDX_DIAL_SERVER_TYPE_UNKNOWN'
};
/**
 * @internal
 *
 * A `BindParams` object holds data required for constructing the query string
 * component of URLs for:
 * 1. Establishing sessions;
 * 2. Sending messages; and
 * 3. Listening for incoming messages.
 *
 * @see {@link Session}
 * @see {@link RPCConnection}
 */
class BindParams {
    constructor(options) {
        _BindParams_instances.add(this);
        this.device = 'LOUNGE_SCREEN';
        this.id = options.deviceId;
        this.obfuscatedGaiaId = '';
        this.name = options.screenName;
        this.app = options.screenApp;
        this.theme = options.theme;
        this.capabilities = 'dsp,mic,dpa,ntb';
        this.cst = 'm';
        this.mdxVersion = 2;
        this.VER = 8;
        this.v = 2;
        this.deviceInfo = {
            ...DEFAULT_DEVICE_INFO,
            brand: options.brand,
            model: options.model
        };
        this.CVER = 1;
        this.zx = __classPrivateFieldGet(this, _BindParams_instances, "m", _BindParams_generateZX).call(this);
        this.t = 1;
        this.RID = __classPrivateFieldGet(this, _BindParams_instances, "m", _BindParams_generateRID).call(this);
        this.AID = 3;
    }
    reset() {
        delete this.SID;
        delete this.loungeIdToken;
        delete this.gsessionid;
        this.RID = __classPrivateFieldGet(this, _BindParams_instances, "m", _BindParams_generateRID).call(this);
        this.AID = 3;
    }
    updateWithMdxContext(context) {
        if (context.deviceId) {
            this.id = context.deviceId;
        }
    }
    updateWithLoungeToken(token) {
        this.loungeIdToken = token.loungeToken;
    }
    updateWithMessage(cmd) {
        if (Array.isArray(cmd)) {
            for (const c of cmd) {
                this.updateWithMessage(c);
            }
            return;
        }
        switch (cmd.name) {
            case 'c':
                // E.g.: [0, ["c", "SID here", "", 8]]
                this.SID = cmd.payload[0];
                break;
            case 'S':
                // E.g.: [1, ["S", "gsessionid here"]]
                this.gsessionid = cmd.payload;
                break;
            default:
        }
        if (cmd.AID) {
            this.AID = Math.max(this.AID, cmd.AID);
        }
    }
    /**
     * Constructs query string for specified action type.
     * @param type - The type of action to be performed.
     * @param AID - Action identifier associated with the action.
     * @returns Query string
     * @throws {@link IncompleteAPIDataError} if data needed to construct the query string is missing.
     */
    toQueryString(type, AID) {
        const missing = [];
        if (!this.loungeIdToken) {
            missing.push('loungeIdToken');
        }
        if (type === 'sendMessage' || type === 'rpc') {
            if (!this.SID) {
                missing.push('SID');
            }
            if (!this.gsessionid) {
                missing.push('gsessionid');
            }
        }
        if (missing.length > 0) {
            throw new IncompleteAPIDataError('Missing data required to construct query string from bind params', missing);
        }
        const commonParams = {
            device: this.device,
            id: this.id,
            obfuscatedGaiaId: this.obfuscatedGaiaId,
            name: this.name,
            app: this.app,
            theme: this.theme,
            capabilities: this.capabilities,
            cst: this.cst,
            mdxVersion: this.mdxVersion,
            loungeIdToken: this.loungeIdToken,
            VER: this.VER,
            v: this.v,
            zx: __classPrivateFieldGet(this, _BindParams_instances, "m", _BindParams_generateZX).call(this),
            t: this.t
        };
        let params;
        switch (type) {
            case 'initSession':
                params = {
                    ...commonParams,
                    deviceInfo: JSON.stringify(this.deviceInfo),
                    RID: this.RID,
                    CVER: this.CVER
                };
                this.RID++;
                break;
            case 'sendMessage':
                this.AID = AID && this.AID ?
                    Math.max(AID, this.AID) : AID || this.AID;
                params = {
                    ...commonParams,
                    deviceInfo: JSON.stringify(this.deviceInfo),
                    SID: this.SID,
                    RID: this.RID,
                    AID: this.AID,
                    gsessionid: this.gsessionid
                };
                if (!AID) {
                    this.AID++;
                }
                this.RID++;
                break;
            case 'rpc':
                params = {
                    ...commonParams,
                    RID: 'rpc',
                    SID: this.SID,
                    CI: 0,
                    AID: this.AID,
                    gsessionid: this.gsessionid,
                    TYPE: 'xmlhttp'
                };
                break;
            default:
        }
        return params ? queryString.stringify(params) : '';
    }
}
_BindParams_instances = new WeakSet(), _BindParams_generateZX = function _BindParams_generateZX() {
    return uuidv4().replace(/-/g, '').substring(0, 12);
}, _BindParams_generateRID = function _BindParams_generateRID() {
    const min = 41000;
    const max = 49999;
    return Math.round(Math.random() * (max - min) + min);
};
export default BindParams;
