// yt-cast-receiver/dist/lib/app/DefaultPlaylistRequestHandler.js
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _DefaultPlaylistRequestHandler_instances, _DefaultPlaylistRequestHandler_innertube, _DefaultPlaylistRequestHandler_innertubeInitialClient, _DefaultPlaylistRequestHandler_innertubeTVClient, _DefaultPlaylistRequestHandler_innertubeMusicClient, _DefaultPlaylistRequestHandler_init, _DefaultPlaylistRequestHandler_createInnertubeEndpoint, _DefaultPlaylistRequestHandler_configureInnertubeContext, _DefaultPlaylistRequestHandler_videoFromInnertubeEndpoint;
import Innertube, * as InnertubeLib from 'youtubei.js';
import { CLIENTS } from '../Constants.js';
import PlaylistRequestHandler from './PlaylistRequestHandler.js';
const TV_CLIENT_INFO = {
    clientName: 'TVHTML5',
    clientVersion: '7.20230405.08.01',
    userAgent: 'Mozilla/5.0 (Windows NT 10.0; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/75.0.3770.142 Safari/537.36; SMART-TV; Tizen 4.0,gzip(gfe)'
};
const MUSIC_CLIENT_INFO = {
    clientName: InnertubeLib.Constants.CLIENTS.YTMUSIC.NAME,
    clientVersion: InnertubeLib.Constants.CLIENTS.YTMUSIC.VERSION
};
/**
 * Default implementation of the `PlaylistRequestHandler` abstract class.
 * Uses [YouTube.js](https://github.com/LuanRT/YouTube.js) for fetching data
 * from YouTube.
 */
class DefaultPlaylistRequestHandler extends PlaylistRequestHandler {
    constructor() {
        super();
        _DefaultPlaylistRequestHandler_instances.add(this);
        _DefaultPlaylistRequestHandler_innertube.set(this, void 0);
        _DefaultPlaylistRequestHandler_innertubeInitialClient.set(this, void 0);
        _DefaultPlaylistRequestHandler_innertubeTVClient.set(this, void 0);
        _DefaultPlaylistRequestHandler_innertubeMusicClient.set(this, void 0);
        __classPrivateFieldSet(this, _DefaultPlaylistRequestHandler_innertube, null, "f");
    }
    async getPreviousNextVideos(target, playlist) {
        if (!__classPrivateFieldGet(this, _DefaultPlaylistRequestHandler_innertube, "f")) {
            await __classPrivateFieldGet(this, _DefaultPlaylistRequestHandler_instances, "m", _DefaultPlaylistRequestHandler_init).call(this);
        }
        if (!__classPrivateFieldGet(this, _DefaultPlaylistRequestHandler_innertube, "f")) {
            throw Error('DefaultPlaylistRequestHandler not initialized');
        }
        this.logger.debug('[yt-cast-receiver] DefaultPlaylistRequestHandler.getPreviousNextVideos:', target);
        /**
         * Mark target as watched to reduce its chance of being returned as autoplay video in future results.
         *
         * Note:
         * By experiment, it takes a bit of time for markWatched() to take effect. For example, let's say video2 is
         * the autoplay result for video1. If, after calling markWatched() for video1, you immediately request
         * next video for video2, there is still a good chance that video1 will be returned as autoplay. But if you
         * do it 10 seconds after markWatched(), then the chance will be greatly reduced. This behavior can be observed
         * even on YouTube website.
         */
        await this.markWatched(target);
        try {
            // We are sending requests as TV client, so need to specify `noClient: true` to remove `client` property from endpoint
            // (really only applies when client is YouTube Music).
            const endpoint = __classPrivateFieldGet(this, _DefaultPlaylistRequestHandler_instances, "m", _DefaultPlaylistRequestHandler_createInnertubeEndpoint).call(this, target, true);
            __classPrivateFieldGet(this, _DefaultPlaylistRequestHandler_instances, "m", _DefaultPlaylistRequestHandler_configureInnertubeContext).call(this, target, 'tv');
            const nextResponse = await __classPrivateFieldGet(this, _DefaultPlaylistRequestHandler_innertube, "f").actions.execute('/next', endpoint.payload);
            const autoplaySet = nextResponse.data?.contents?.singleColumnWatchNextResults?.autoplay?.autoplay
                ?.sets?.[0];
            const previousRendererEndpoint = autoplaySet?.previousVideoRenderer?.autoplayEndpointRenderer?.endpoint;
            const nextRendererEndpoint = autoplaySet?.nextVideoRenderer?.autoplayEndpointRenderer?.endpoint;
            const autoplayRendererEndpoint = autoplaySet?.autoplayVideoRenderer?.mdxAutoplayVideoRenderer?.navigationEndpoint;
            const parsedEndpoints = {
                previous: previousRendererEndpoint ? new InnertubeLib.YTNodes.NavigationEndpoint(previousRendererEndpoint) : null,
                next: nextRendererEndpoint ? new InnertubeLib.YTNodes.NavigationEndpoint(nextRendererEndpoint) : null
            };
            let nextIsAutoplay = false;
            if (!parsedEndpoints.next && autoplayRendererEndpoint) {
                parsedEndpoints.next = new InnertubeLib.YTNodes.NavigationEndpoint(autoplayRendererEndpoint);
                nextIsAutoplay = true;
            }
            this.logger.debug(`[yt-cast-receiver] Previous / Next endpoints for video Id: ${target.id}`, parsedEndpoints);
            const result = {
                previous: parsedEndpoints.previous ? __classPrivateFieldGet(this, _DefaultPlaylistRequestHandler_instances, "m", _DefaultPlaylistRequestHandler_videoFromInnertubeEndpoint).call(this, parsedEndpoints.previous, target.client) : null,
                next: parsedEndpoints.next ? __classPrivateFieldGet(this, _DefaultPlaylistRequestHandler_instances, "m", _DefaultPlaylistRequestHandler_videoFromInnertubeEndpoint).call(this, parsedEndpoints.next, target.client) : null
            };
            this.logger.debug(`[yt-cast-receiver] Previous / Next videos for video Id: ${target.id}`, result);
            if (nextIsAutoplay && result.next?.id && playlist.videoIds.includes(result.next.id)) {
                this.logger.warn(`[yt-cast-receiver] Autoplay video ${result.next.id} already exists in playlist.`);
            }
            return result;
        }
        catch (error) {
            this.logger.error(`[yt-cast-receiver] Failed to get autoplay video for video Id: ${target.id}`, error);
            return { previous: null, next: null };
        }
    }
    async markWatched(video, run = 1) {
        if (!__classPrivateFieldGet(this, _DefaultPlaylistRequestHandler_innertube, "f")) {
            throw Error('DefaultPlaylistRequestHandler not initialized.');
        }
        /**
         * MarkWatched() involves 2 runs: once with playlist info intact, and once without.
         * However, 2nd run will be skipped if video is private because without playlist info, player response won't contain the
         * Playback tracking URL needed to mark video as watched. Presumably, this is because a private video cannot exist in public space.
         */
        this.logger.debug(`[yt-cast-receiver] Marking video ${video.id} as watched ${run === 1 ? '(1st run)' : '(2nd run - without playlist info)'}...`);
        const isMusic = video.client === CLIENTS.YTMUSIC;
        __classPrivateFieldGet(this, _DefaultPlaylistRequestHandler_instances, "m", _DefaultPlaylistRequestHandler_configureInnertubeContext).call(this, video, isMusic ? 'music' : 'initial');
        const endpoint = __classPrivateFieldGet(this, _DefaultPlaylistRequestHandler_instances, "m", _DefaultPlaylistRequestHandler_createInnertubeEndpoint).call(this, run === 1 ? video : { ...video, context: undefined });
        const payload = endpoint.payload;
        try {
            const playerResponse = await __classPrivateFieldGet(this, _DefaultPlaylistRequestHandler_innertube, "f").actions.execute('/player', payload);
            const playbackUrl = playerResponse.data?.playbackTracking?.videostatsPlaybackUrl?.baseUrl;
            if (playbackUrl) {
                let referrer = `https://www.youtube.com/tv#/watch?v=${video.id}`;
                if (video.context?.playlistId) {
                    referrer += `&list=${video.context.playlistId}`;
                }
                const clientInfo = isMusic ? MUSIC_CLIENT_INFO : TV_CLIENT_INFO;
                const params = {
                    cpn: InnertubeLib.Utils.generateRandomString(16),
                    fmt: 251,
                    rtn: 0,
                    rt: 0,
                    c: clientInfo.clientName,
                    cver: clientInfo.clientVersion,
                    ctheme: isMusic ? 'MUSIC' : 'CLASSIC',
                    cplayer: 'UNIPLAYER',
                    ctrl: 'mdx-pair',
                    referrer
                };
                if (video.context?.ctt) {
                    params.ctt = video.context.ctt;
                    params.cttype = 'vvt';
                }
                await __classPrivateFieldGet(this, _DefaultPlaylistRequestHandler_innertube, "f").session.actions.stats(playbackUrl, {
                    client_name: clientInfo.clientName,
                    client_version: clientInfo.clientVersion
                }, params);
                if (run === 2) {
                    return;
                }
                if (playerResponse.data?.videoDetails?.isPrivate) {
                    this.logger.debug('[yt-cast-receiver] DefaultPlaylistRequestHandler skipping 2nd run of markWatched(): video is private and cannot be mark as watched without playlist info.');
                    return;
                }
                return await this.markWatched(video, 2);
            }
            throw Error('No playback tracking URL found');
        }
        catch (error) {
            this.logger.warn(`[yt-cast-receiver] DefaultPlaylistRequestHandler failed to mark video ${video.id} as watched:`, error);
        }
    }
    // Overrides
    reset() {
        __classPrivateFieldSet(this, _DefaultPlaylistRequestHandler_innertube, null, "f");
    }
}
_DefaultPlaylistRequestHandler_innertube = new WeakMap(), _DefaultPlaylistRequestHandler_innertubeInitialClient = new WeakMap(), _DefaultPlaylistRequestHandler_innertubeTVClient = new WeakMap(), _DefaultPlaylistRequestHandler_innertubeMusicClient = new WeakMap(), _DefaultPlaylistRequestHandler_instances = new WeakSet(), _DefaultPlaylistRequestHandler_init = async function _DefaultPlaylistRequestHandler_init() {
    if (!__classPrivateFieldGet(this, _DefaultPlaylistRequestHandler_innertube, "f")) {
        // https://github.com/LuanRT/YouTube.js/issues/1043
        __classPrivateFieldSet(this, _DefaultPlaylistRequestHandler_innertube, await Innertube.create(), "f");
        __classPrivateFieldSet(this, _DefaultPlaylistRequestHandler_innertubeInitialClient, { ...__classPrivateFieldGet(this, _DefaultPlaylistRequestHandler_innertube, "f").session.context.client }, "f");
        __classPrivateFieldSet(this, _DefaultPlaylistRequestHandler_innertubeTVClient, {
            ...__classPrivateFieldGet(this, _DefaultPlaylistRequestHandler_innertube, "f").session.context.client,
            ...TV_CLIENT_INFO
        }, "f");
        __classPrivateFieldSet(this, _DefaultPlaylistRequestHandler_innertubeMusicClient, {
            ...__classPrivateFieldGet(this, _DefaultPlaylistRequestHandler_innertube, "f").session.context.client,
            ...MUSIC_CLIENT_INFO
        }, "f");
    }
}, _DefaultPlaylistRequestHandler_createInnertubeEndpoint = function _DefaultPlaylistRequestHandler_createInnertubeEndpoint(video, noClient = false) {
    if (!__classPrivateFieldGet(this, _DefaultPlaylistRequestHandler_innertube, "f")) {
        throw Error('DefaultPlaylistRequestHandler not initialized.');
    }
    const endpoint = new InnertubeLib.YTNodes.NavigationEndpoint({});
    endpoint.payload = {
        videoId: video.id,
        enableMdxAutoplay: true,
        isMdxPlayback: true
    };
    if (video.context?.playlistId) {
        endpoint.payload.playlistId = video.context.playlistId;
    }
    if (video.context?.params) {
        endpoint.payload.params = video.context.params;
    }
    if (video.context?.index !== undefined) {
        endpoint.payload.playlistIndex = video.context.index;
    }
    if (video.client === CLIENTS.YTMUSIC && !noClient) {
        endpoint.payload.client = 'YTMUSIC';
    }
    return endpoint;
}, _DefaultPlaylistRequestHandler_configureInnertubeContext = function _DefaultPlaylistRequestHandler_configureInnertubeContext(video, type) {
    if (!__classPrivateFieldGet(this, _DefaultPlaylistRequestHandler_innertube, "f")) {
        throw Error('DefaultPlaylistRequestHandler not initialized.');
    }
    switch (type) {
        case 'tv':
            __classPrivateFieldGet(this, _DefaultPlaylistRequestHandler_innertube, "f").session.context.client = __classPrivateFieldGet(this, _DefaultPlaylistRequestHandler_innertubeTVClient, "f");
            break;
        case 'music':
            __classPrivateFieldGet(this, _DefaultPlaylistRequestHandler_innertube, "f").session.context.client = __classPrivateFieldGet(this, _DefaultPlaylistRequestHandler_innertubeMusicClient, "f");
            break;
        default:
            __classPrivateFieldGet(this, _DefaultPlaylistRequestHandler_innertube, "f").session.context.client = __classPrivateFieldGet(this, _DefaultPlaylistRequestHandler_innertubeInitialClient, "f");
    }
    if (video.context?.ctt) {
        __classPrivateFieldGet(this, _DefaultPlaylistRequestHandler_innertube, "f").session.context.user = {
            enableSafetyMode: false,
            lockedSafetyMode: false,
            credentialTransferTokens: [
                {
                    'scope': 'VIDEO',
                    'token': video.context?.ctt
                }
            ]
        };
    }
    else {
        delete __classPrivateFieldGet(this, _DefaultPlaylistRequestHandler_innertube, "f").session.context.user?.credentialTransferTokens;
    }
}, _DefaultPlaylistRequestHandler_videoFromInnertubeEndpoint = function _DefaultPlaylistRequestHandler_videoFromInnertubeEndpoint(endpoint, client) {
    if (!endpoint.payload?.videoId) {
        return null;
    }
    const result = {
        id: endpoint.payload.videoId,
        client
    };
    const context = {};
    if (endpoint.payload.index !== undefined) {
        context.index = endpoint.payload.index;
    }
    if (endpoint.payload.playlistId) {
        context.playlistId = endpoint.payload.playlistId;
    }
    if (endpoint.payload.params) {
        context.params = endpoint.payload.params;
    }
    if (endpoint.payload.watchEndpointSupportedAuthorizationTokenConfig?.videoAuthorizationToken?.credentialTransferTokens?.[0]?.token) {
        context.ctt = endpoint.payload.watchEndpointSupportedAuthorizationTokenConfig?.videoAuthorizationToken?.credentialTransferTokens?.[0]?.token;
    }
    if (Object.keys(context).length > 0) {
        result.context = context;
    }
    return result;
};
export default DefaultPlaylistRequestHandler;
