// yt-cast-receiver/dist/lib/app/PairingCodeRequestService.js
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _PairingCodeRequestService_instances, _PairingCodeRequestService_screen, _PairingCodeRequestService_bindParams, _PairingCodeRequestService_status, _PairingCodeRequestService_refreshTimer, _PairingCodeRequestService_clearRefreshTimer, _PairingCodeRequestService_startRefreshTimer, _PairingCodeRequestService_getCodeAndEmit;
import EventEmitter from 'events';
import { ConnectionError, DataError } from '../utils/Errors.js';
import { STATUSES, URLS } from '../Constants.js';
const REFRESH_INTERVAL = 300000; // 5 minutes
const REFRESH_INTERVAL_SHORT = 30000; // 30 seconds - when API not ready
/**
 * Fetches pairing code for manual pairing (aka 'Link with TV code').
 * A pairing code is refreshed every 5 minutes. Results are returned through
 * the `response` event.
 *
 * ### Usage
 *
 * ```
 * const service = receiver.getPairingCodeRequestService();
 *
 * service.on('request', () => {...});  // Event when request is being made
 * service.on('response', (code) => {...}); // Event when code is obtained
 * service.on('error', (error) => {...}); // Event when error occurs
 *
 * service.start();
 * ...
 *
 * service.stop();
 * ```
 *
 * Note that the service stops on `error` event.
 */
class PairingCodeRequestService extends EventEmitter {
    /**
     * @internal
     */
    constructor(screen, bindParams) {
        super();
        _PairingCodeRequestService_instances.add(this);
        _PairingCodeRequestService_screen.set(this, void 0);
        _PairingCodeRequestService_bindParams.set(this, void 0);
        _PairingCodeRequestService_status.set(this, void 0);
        _PairingCodeRequestService_refreshTimer.set(this, void 0);
        __classPrivateFieldSet(this, _PairingCodeRequestService_bindParams, bindParams, "f");
        __classPrivateFieldSet(this, _PairingCodeRequestService_screen, screen, "f");
        __classPrivateFieldSet(this, _PairingCodeRequestService_refreshTimer, null, "f");
        __classPrivateFieldSet(this, _PairingCodeRequestService_status, STATUSES.STOPPED, "f");
    }
    /**
     * Starts the service.
     */
    start() {
        if (__classPrivateFieldGet(this, _PairingCodeRequestService_status, "f") === STATUSES.RUNNING) {
            return;
        }
        __classPrivateFieldSet(this, _PairingCodeRequestService_status, STATUSES.RUNNING, "f");
        void __classPrivateFieldGet(this, _PairingCodeRequestService_instances, "m", _PairingCodeRequestService_getCodeAndEmit).call(this);
    }
    /**
     * Stops the service.
     */
    stop() {
        if (__classPrivateFieldGet(this, _PairingCodeRequestService_status, "f") === STATUSES.STOPPED) {
            return;
        }
        __classPrivateFieldGet(this, _PairingCodeRequestService_instances, "m", _PairingCodeRequestService_clearRefreshTimer).call(this);
        __classPrivateFieldSet(this, _PairingCodeRequestService_status, STATUSES.STOPPED, "f");
    }
    on(event, listener) {
        super.on(event, listener);
        return this;
    }
    /** Service status */
    get status() {
        return __classPrivateFieldGet(this, _PairingCodeRequestService_status, "f");
    }
}
_PairingCodeRequestService_screen = new WeakMap(), _PairingCodeRequestService_bindParams = new WeakMap(), _PairingCodeRequestService_status = new WeakMap(), _PairingCodeRequestService_refreshTimer = new WeakMap(), _PairingCodeRequestService_instances = new WeakSet(), _PairingCodeRequestService_clearRefreshTimer = function _PairingCodeRequestService_clearRefreshTimer() {
    if (__classPrivateFieldGet(this, _PairingCodeRequestService_refreshTimer, "f")) {
        clearTimeout(__classPrivateFieldGet(this, _PairingCodeRequestService_refreshTimer, "f"));
        __classPrivateFieldSet(this, _PairingCodeRequestService_refreshTimer, null, "f");
    }
}, _PairingCodeRequestService_startRefreshTimer = function _PairingCodeRequestService_startRefreshTimer(interval = REFRESH_INTERVAL) {
    __classPrivateFieldGet(this, _PairingCodeRequestService_instances, "m", _PairingCodeRequestService_clearRefreshTimer).call(this);
    __classPrivateFieldSet(this, _PairingCodeRequestService_refreshTimer, setTimeout(() => {
        void __classPrivateFieldGet(this, _PairingCodeRequestService_instances, "m", _PairingCodeRequestService_getCodeAndEmit).call(this);
    }, interval), "f");
}, _PairingCodeRequestService_getCodeAndEmit = async function _PairingCodeRequestService_getCodeAndEmit() {
    __classPrivateFieldGet(this, _PairingCodeRequestService_instances, "m", _PairingCodeRequestService_clearRefreshTimer).call(this);
    if (!__classPrivateFieldGet(this, _PairingCodeRequestService_bindParams, "f").loungeIdToken || !__classPrivateFieldGet(this, _PairingCodeRequestService_screen, "f").id) {
        // API not ready
        __classPrivateFieldGet(this, _PairingCodeRequestService_instances, "m", _PairingCodeRequestService_startRefreshTimer).call(this, REFRESH_INTERVAL_SHORT);
        return;
    }
    const params = new URLSearchParams({
        'access_type': 'permanent',
        'app': __classPrivateFieldGet(this, _PairingCodeRequestService_screen, "f").app,
        'lounge_token': __classPrivateFieldGet(this, _PairingCodeRequestService_bindParams, "f").loungeIdToken,
        'screen_id': __classPrivateFieldGet(this, _PairingCodeRequestService_screen, "f").id,
        'screen_name': __classPrivateFieldGet(this, _PairingCodeRequestService_screen, "f").name,
        'device_id': __classPrivateFieldGet(this, _PairingCodeRequestService_bindParams, "f").id
    });
    const url = URLS.GET_PAIRING_CODE;
    let response;
    let error;
    try {
        this.emit('request');
        response = await fetch(url, {
            method: 'POST',
            body: params
        });
    }
    catch (err) {
        error = new ConnectionError('Connection error in fetching pairing code', url, err);
        response = null;
    }
    if (response) {
        try {
            const code = await response.text();
            this.emit('response', code);
            __classPrivateFieldGet(this, _PairingCodeRequestService_instances, "m", _PairingCodeRequestService_startRefreshTimer).call(this);
        }
        catch (err) {
            error = new DataError('Pairing code data error', err);
        }
    }
    if (error) {
        this.stop();
        this.emit('error', error);
    }
};
export default PairingCodeRequestService;
