// yt-cast-receiver/dist/lib/app/Playlist.js
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _Playlist_instances, _Playlist_id, _Playlist_videoIds, _Playlist_current, _Playlist_previous, _Playlist_next, _Playlist_autoplayMode, _Playlist_logger, _Playlist_requestHandler, _Playlist_previousNextAbortController, _Playlist_abortRefreshPreviousNext, _Playlist_refreshPreviousNext;
import EventEmitter from 'events';
import { AUTOPLAY_MODES } from '../Constants.js';
export const PLAYLIST_EVENT_TYPES = {
    VIDEO_SELECTED: 'videoSelected',
    VIDEO_ADDED: 'videoAdded',
    VIDEO_REMOVED: 'videoRemoved',
    PLAYLIST_SET: 'playlistSet',
    PLAYLIST_ADDED: 'playlistAdded',
    PLAYLIST_CLEARED: 'playlistCleared',
    PLAYLIST_UPDATED: 'playlistUpdated'
};
/**
 * Representation of the player queue.
 */
class Playlist extends EventEmitter {
    /** @internal */
    constructor() {
        super();
        _Playlist_instances.add(this);
        _Playlist_id.set(this, void 0);
        _Playlist_videoIds.set(this, void 0);
        _Playlist_current.set(this, void 0);
        _Playlist_previous.set(this, void 0);
        _Playlist_next.set(this, void 0);
        _Playlist_autoplayMode.set(this, void 0);
        _Playlist_logger.set(this, void 0);
        _Playlist_requestHandler.set(this, void 0);
        _Playlist_previousNextAbortController.set(this, void 0);
        __classPrivateFieldSet(this, _Playlist_id, null, "f");
        __classPrivateFieldSet(this, _Playlist_videoIds, [], "f");
        __classPrivateFieldSet(this, _Playlist_current, null, "f");
        __classPrivateFieldSet(this, _Playlist_previous, null, "f");
        __classPrivateFieldSet(this, _Playlist_autoplayMode, AUTOPLAY_MODES.UNSUPPORTED, "f");
        __classPrivateFieldSet(this, _Playlist_previousNextAbortController, null, "f");
    }
    /** @internal */
    setLogger(logger) {
        __classPrivateFieldSet(this, _Playlist_logger, logger, "f");
    }
    /** @internal */
    setRequestHandler(value) {
        __classPrivateFieldSet(this, _Playlist_requestHandler, value, "f");
    }
    /** @internal */
    async setAutoplayMode(value) {
        const oldMode = __classPrivateFieldGet(this, _Playlist_autoplayMode, "f");
        __classPrivateFieldSet(this, _Playlist_autoplayMode, value, "f");
        if (oldMode !== value) {
            await __classPrivateFieldGet(this, _Playlist_instances, "m", _Playlist_refreshPreviousNext).call(this);
            this.emit('autoplayModeChange', oldMode, value);
        }
    }
    /**
     * @internal
     *
     * Updates the playlist with payload of 'setPlaylist' or 'updatePlaylist' message.
     * @param data - 'setPlaylist' or 'updatePlaylist' `Message` object.
     */
    async updateByMessage(message, client) {
        if (message.name !== 'setPlaylist' && message.name !== 'updatePlaylist') {
            return;
        }
        const data = message.payload;
        if (!data.listId) {
            __classPrivateFieldSet(this, _Playlist_id, null, "f");
            __classPrivateFieldSet(this, _Playlist_current, null, "f");
            __classPrivateFieldSet(this, _Playlist_next, null, "f");
            __classPrivateFieldSet(this, _Playlist_previous, null, "f");
            __classPrivateFieldSet(this, _Playlist_videoIds, [], "f");
            this.emit('playlistCleared', { type: 'playlistCleared' });
            return;
        }
        __classPrivateFieldSet(this, _Playlist_id, data.listId, "f");
        __classPrivateFieldSet(this, _Playlist_videoIds, data.videoIds ? data.videoIds.split(',') : [], "f");
        if (message.name === 'setPlaylist') {
            if (data.currentIndex && data.videoId) {
                const context = {
                    playlistId: data.listId,
                    index: parseInt(data.currentIndex, 10)
                };
                if (data.ctt) {
                    context.ctt = data.ctt;
                }
                if (data.params) {
                    context.params = data.params;
                }
                __classPrivateFieldSet(this, _Playlist_current, {
                    id: data.videoId,
                    client,
                    context
                }, "f");
            }
            else {
                __classPrivateFieldSet(this, _Playlist_current, null, "f");
            }
        }
        else if (__classPrivateFieldGet(this, _Playlist_current, "f")) { // 'updatePlaylist'
            const index = __classPrivateFieldGet(this, _Playlist_videoIds, "f").findIndex((v) => v === __classPrivateFieldGet(this, _Playlist_current, "f")?.id);
            if (index >= 0) {
                if (!__classPrivateFieldGet(this, _Playlist_current, "f").context) {
                    __classPrivateFieldGet(this, _Playlist_current, "f").context = {};
                }
                __classPrivateFieldGet(this, _Playlist_current, "f").context.playlistId = data.listId;
                __classPrivateFieldGet(this, _Playlist_current, "f").context.index = index;
                if (data.params) {
                    __classPrivateFieldGet(this, _Playlist_current, "f").context.params = data.params;
                }
                else {
                    delete __classPrivateFieldGet(this, _Playlist_current, "f").context.params;
                }
            }
            else {
                __classPrivateFieldSet(this, _Playlist_current, null, "f");
            }
        }
        await __classPrivateFieldGet(this, _Playlist_instances, "m", _Playlist_refreshPreviousNext).call(this);
        let event = null;
        if (data.eventDetails) {
            try {
                const eventDetails = JSON.parse(data.eventDetails);
                const eventType = eventDetails.eventType;
                let emitEventType;
                switch (eventType) {
                    case 'PLAYLIST_ADDED':
                        emitEventType = 'playlistAdded';
                        break;
                    case 'PLAYLIST_CLEARED':
                        emitEventType = 'playlistCleared';
                        break;
                    case 'PLAYLIST_SET':
                        emitEventType = 'playlistSet';
                        break;
                    case 'VIDEO_ADDED':
                        emitEventType = 'videoAdded';
                        break;
                    case 'VIDEO_REMOVED':
                        emitEventType = 'videoRemoved';
                        break;
                    case 'VIDEO_SELECTED':
                        emitEventType = 'videoSelected';
                        break;
                    default:
                        emitEventType = null;
                }
                if (emitEventType) {
                    event = {
                        type: emitEventType
                    };
                    if (eventDetails.user) {
                        event.user = {
                            name: eventDetails.user,
                            thumbnail: eventDetails.userAvatarUri
                        };
                    }
                    if (eventDetails.videoId) {
                        event.videoId = eventDetails.videoId;
                    }
                    if (eventDetails.videoIds) {
                        event.videoIds = eventDetails.videoIds;
                    }
                }
            }
            catch (error) {
                __classPrivateFieldGet(this, _Playlist_logger, "f").error('[yt-cast-receiver] Failed to parse playlist eventDetails:', data.eventDetails, error);
                event = null;
            }
        }
        if (!event) {
            const emitEventType = message.name === 'setPlaylist' ? 'playlistSet' : 'playlistUpdated';
            event = {
                type: emitEventType,
                videoIds: __classPrivateFieldGet(this, _Playlist_videoIds, "f")
            };
        }
        this.emit(event.type, event);
    }
    /** @internal */
    reset() {
        __classPrivateFieldSet(this, _Playlist_id, null, "f");
        __classPrivateFieldSet(this, _Playlist_videoIds, [], "f");
        __classPrivateFieldSet(this, _Playlist_current, null, "f");
        __classPrivateFieldSet(this, _Playlist_previous, null, "f");
        __classPrivateFieldSet(this, _Playlist_autoplayMode, AUTOPLAY_MODES.UNSUPPORTED, "f");
        __classPrivateFieldGet(this, _Playlist_requestHandler, "f").reset();
    }
    /**
     * Id of the playlist.
     */
    get id() {
        return __classPrivateFieldGet(this, _Playlist_id, "f");
    }
    /**
     * The Ids of the videos in the playlist.
     */
    get videoIds() {
        return __classPrivateFieldGet(this, _Playlist_videoIds, "f");
    }
    /**
     * The number of videos in the playlist.
     */
    get length() {
        return __classPrivateFieldGet(this, _Playlist_videoIds, "f").length;
    }
    /** @internal */
    async previous() {
        if (!__classPrivateFieldGet(this, _Playlist_previous, "f")) {
            return null;
        }
        __classPrivateFieldSet(this, _Playlist_current, __classPrivateFieldGet(this, _Playlist_previous, "f"), "f");
        await __classPrivateFieldGet(this, _Playlist_instances, "m", _Playlist_refreshPreviousNext).call(this);
        return __classPrivateFieldGet(this, _Playlist_current, "f");
    }
    /** @internal */
    async next() {
        if (!this.hasNext) {
            return null;
        }
        __classPrivateFieldSet(this, _Playlist_current, __classPrivateFieldGet(this, _Playlist_next, "f"), "f");
        await __classPrivateFieldGet(this, _Playlist_instances, "m", _Playlist_refreshPreviousNext).call(this);
        return __classPrivateFieldGet(this, _Playlist_current, "f");
    }
    getState() {
        return {
            id: this.id,
            videoIds: this.videoIds,
            previous: __classPrivateFieldGet(this, _Playlist_previous, "f"),
            current: this.current,
            next: this.hasNext ? __classPrivateFieldGet(this, _Playlist_next, "f") : null,
            autoplay: this.autoplay
        };
    }
    /** @internal */
    setAsCurrent(video) {
        if (__classPrivateFieldGet(this, _Playlist_current, "f")?.id === video.id) {
            return;
        }
        if (this.autoplay?.id === video.id) {
            __classPrivateFieldSet(this, _Playlist_next, null, "f");
        }
        __classPrivateFieldSet(this, _Playlist_current, video, "f");
    }
    get autoplay() {
        if (this.isLast) {
            return __classPrivateFieldGet(this, _Playlist_next, "f");
        }
        return null;
    }
    get current() {
        return __classPrivateFieldGet(this, _Playlist_current, "f");
    }
    get autoplayMode() {
        return __classPrivateFieldGet(this, _Playlist_autoplayMode, "f");
    }
    get isLast() {
        const currentIndex = __classPrivateFieldGet(this, _Playlist_current, "f")?.context?.index !== undefined ? __classPrivateFieldGet(this, _Playlist_current, "f").context.index : -1;
        return currentIndex < 0 || currentIndex === this.length - 1;
    }
    get hasPrevious() {
        return !!__classPrivateFieldGet(this, _Playlist_previous, "f");
    }
    get hasNext() {
        if (!__classPrivateFieldGet(this, _Playlist_next, "f") || this.isLast) {
            return false;
        }
        return true;
    }
    get isUpdating() {
        return !!__classPrivateFieldGet(this, _Playlist_previousNextAbortController, "f");
    }
    /**
     * @internal
     */
    get requestHandler() {
        return __classPrivateFieldGet(this, _Playlist_requestHandler, "f");
    }
    on(event, listener) {
        super.on(event, listener);
        return this;
    }
}
_Playlist_id = new WeakMap(), _Playlist_videoIds = new WeakMap(), _Playlist_current = new WeakMap(), _Playlist_previous = new WeakMap(), _Playlist_next = new WeakMap(), _Playlist_autoplayMode = new WeakMap(), _Playlist_logger = new WeakMap(), _Playlist_requestHandler = new WeakMap(), _Playlist_previousNextAbortController = new WeakMap(), _Playlist_instances = new WeakSet(), _Playlist_abortRefreshPreviousNext = function _Playlist_abortRefreshPreviousNext() {
    if (__classPrivateFieldGet(this, _Playlist_previousNextAbortController, "f")) {
        __classPrivateFieldGet(this, _Playlist_previousNextAbortController, "f").abort();
        __classPrivateFieldSet(this, _Playlist_previousNextAbortController, null, "f");
    }
}, _Playlist_refreshPreviousNext = async function _Playlist_refreshPreviousNext() {
    __classPrivateFieldGet(this, _Playlist_instances, "m", _Playlist_abortRefreshPreviousNext).call(this);
    __classPrivateFieldSet(this, _Playlist_previous, null, "f");
    __classPrivateFieldSet(this, _Playlist_next, null, "f");
    const currentIndex = __classPrivateFieldGet(this, _Playlist_current, "f")?.context?.index !== undefined ? __classPrivateFieldGet(this, _Playlist_current, "f").context.index : -1;
    if (currentIndex >= 0) {
        let nav = null;
        if (__classPrivateFieldGet(this, _Playlist_current, "f")) {
            __classPrivateFieldSet(this, _Playlist_previousNextAbortController, new AbortController(), "f");
            try {
                nav = await __classPrivateFieldGet(this, _Playlist_requestHandler, "f").getPreviousNextVideosAbortable(__classPrivateFieldGet(this, _Playlist_current, "f"), this, __classPrivateFieldGet(this, _Playlist_previousNextAbortController, "f").signal);
            }
            catch (error) {
                if (error.name === 'AbortError') {
                    return;
                }
                throw error;
            }
            finally {
                __classPrivateFieldSet(this, _Playlist_previousNextAbortController, null, "f");
            }
        }
        __classPrivateFieldSet(this, _Playlist_previous, nav?.previous || null, "f");
        if (!this.isLast || __classPrivateFieldGet(this, _Playlist_autoplayMode, "f") === AUTOPLAY_MODES.ENABLED) {
            __classPrivateFieldSet(this, _Playlist_next, nav?.next || null, "f");
        }
    }
};
export default Playlist;
