// yt-cast-receiver/dist/lib/app/PlaylistRequestHandler.js
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _PlaylistRequestHandler_logger;
/**
 * Handles requests made by a `Playlist` instance.
 */
class PlaylistRequestHandler {
    constructor() {
        _PlaylistRequestHandler_logger.set(this, void 0);
    }
    setLogger(logger) {
        __classPrivateFieldSet(this, _PlaylistRequestHandler_logger, logger, "f");
    }
    async getPreviousNextVideosAbortable(target, playlist, abortSignal) {
        const checkAbortSignal = () => {
            if (abortSignal.aborted) {
                const msg = `PlaylistRequestHandler.getPreviousNextVideos() aborted for video Id: ${target.id}`;
                __classPrivateFieldGet(this, _PlaylistRequestHandler_logger, "f").debug(`[yt-cast-receiver] ${msg}.`);
                const abortError = Error(msg);
                abortError.name = 'AbortError';
                throw abortError;
            }
        };
        checkAbortSignal();
        const result = await this.getPreviousNextVideos(target, playlist);
        checkAbortSignal();
        return result;
    }
    /**
     * Resets the handler to its initial state. By default, this method does nothing.
     * Implementations shall override this method if need be.
     */
    reset() {
        // Do nothing
    }
    get logger() {
        return __classPrivateFieldGet(this, _PlaylistRequestHandler_logger, "f");
    }
}
_PlaylistRequestHandler_logger = new WeakMap();
export default PlaylistRequestHandler;
