// yt-cast-receiver/dist/lib/app/RPCConnection.js
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _RPCConnection_instances, _RPCConnection_bindParams, _RPCConnection_logger, _RPCConnection_status, _RPCConnection_abortController, _RPCConnection_reader, _RPCConnection_doConnect, _RPCConnection_handleDisconnect;
import EventEmitter from 'events';
import LineByLineReader from 'line-by-line';
import { Readable } from 'stream';
import Message from './Message.js';
import { AbortError, BadResponseError, ConnectionError } from '../utils/Errors.js';
import { URLS } from '../Constants.js';
const MAX_RETRIES = 3;
/**
 * @internal
 *
 * Connects to YouTube 'bind' URL and listens for messages.
 * Incoming messages are parsed into {@link Message} objects and returned through events.
 *
 * When the underlying connection is ended by remote, which is inevitable under normal circumstance,
 * the `RPCConnection` instance will automatically reconnect.
 */
class RPCConnection extends EventEmitter {
    constructor(options) {
        super();
        _RPCConnection_instances.add(this);
        _RPCConnection_bindParams.set(this, void 0);
        _RPCConnection_logger.set(this, void 0);
        _RPCConnection_status.set(this, void 0);
        _RPCConnection_abortController.set(this, void 0);
        _RPCConnection_reader.set(this, void 0);
        __classPrivateFieldSet(this, _RPCConnection_bindParams, options.bindParams, "f");
        __classPrivateFieldSet(this, _RPCConnection_logger, options.logger, "f");
        __classPrivateFieldSet(this, _RPCConnection_status, 'disconnected', "f");
        __classPrivateFieldSet(this, _RPCConnection_abortController, null, "f");
        __classPrivateFieldSet(this, _RPCConnection_reader, null, "f");
    }
    async connect() {
        if (__classPrivateFieldGet(this, _RPCConnection_status, "f") === 'disconnected') {
            return __classPrivateFieldGet(this, _RPCConnection_instances, "m", _RPCConnection_doConnect).call(this);
        }
    }
    close() {
        if (__classPrivateFieldGet(this, _RPCConnection_status, "f") === 'connected' || __classPrivateFieldGet(this, _RPCConnection_status, "f") === 'connecting' || __classPrivateFieldGet(this, _RPCConnection_status, "f") === 'reconnecting') {
            __classPrivateFieldGet(this, _RPCConnection_logger, "f").debug('[yt-cast-receiver] Closing RPC connection...');
            __classPrivateFieldSet(this, _RPCConnection_status, 'disconnecting', "f");
            if (__classPrivateFieldGet(this, _RPCConnection_abortController, "f")) {
                __classPrivateFieldGet(this, _RPCConnection_abortController, "f").abort();
            }
            if (__classPrivateFieldGet(this, _RPCConnection_reader, "f")) {
                __classPrivateFieldGet(this, _RPCConnection_reader, "f").close();
            }
        }
    }
    on(event, listener) {
        super.on(event, listener);
        return this;
    }
}
_RPCConnection_bindParams = new WeakMap(), _RPCConnection_logger = new WeakMap(), _RPCConnection_status = new WeakMap(), _RPCConnection_abortController = new WeakMap(), _RPCConnection_reader = new WeakMap(), _RPCConnection_instances = new WeakSet(), _RPCConnection_doConnect = async function _RPCConnection_doConnect(isReconnect = false, retry = 0) {
    __classPrivateFieldSet(this, _RPCConnection_status, isReconnect ? 'reconnecting' : 'connecting', "f");
    const url = `${URLS.BIND}?${__classPrivateFieldGet(this, _RPCConnection_bindParams, "f").toQueryString('rpc')}`;
    __classPrivateFieldGet(this, _RPCConnection_logger, "f").debug(`[yt-cast-receiver] Connecting to RPC URL: ${url}`);
    __classPrivateFieldSet(this, _RPCConnection_abortController, new AbortController(), "f");
    let response;
    try {
        response = await fetch(url, { signal: __classPrivateFieldGet(this, _RPCConnection_abortController, "f").signal });
    }
    catch (error) {
        if (error.name === 'AbortError') {
            __classPrivateFieldGet(this, _RPCConnection_logger, "f").debug('[yt-cast-receiver] RPC connection request aborted.');
            __classPrivateFieldSet(this, _RPCConnection_status, 'disconnected', "f");
            throw new AbortError('RPC connection request aborted', url);
        }
        __classPrivateFieldGet(this, _RPCConnection_logger, "f").error('[yt-cast-receiver] RPC connection error:', error);
        retry++;
        if (retry <= MAX_RETRIES) {
            __classPrivateFieldGet(this, _RPCConnection_logger, "f").error(`[yt-cast-receiver] Retrying ${retry} / ${MAX_RETRIES}`);
            return await __classPrivateFieldGet(this, _RPCConnection_instances, "m", _RPCConnection_doConnect).call(this, isReconnect, retry);
        }
        __classPrivateFieldSet(this, _RPCConnection_status, 'disconnected', "f");
        __classPrivateFieldGet(this, _RPCConnection_logger, "f").error('[yt-cast-receiver] Max retries reached. Giving up...');
        throw new ConnectionError('RPC connection error', url, error);
    }
    finally {
        __classPrivateFieldSet(this, _RPCConnection_abortController, null, "f");
    }
    if (response.ok && response.body) {
        __classPrivateFieldGet(this, _RPCConnection_logger, "f").debug('[yt-cast-receiver] RPC connection established.');
        __classPrivateFieldSet(this, _RPCConnection_status, 'connected', "f");
        const readable = Readable.fromWeb(response.body);
        __classPrivateFieldSet(this, _RPCConnection_reader, new LineByLineReader(readable, {
            encoding: 'utf8',
            skipEmptyLines: true
        }), "f");
        __classPrivateFieldGet(this, _RPCConnection_reader, "f").on('line', (line) => {
            if (__classPrivateFieldGet(this, _RPCConnection_status, "f") === 'connected') {
                const messages = Message.parseIncoming(line);
                if (messages.length > 0) {
                    this.emit('messages', messages);
                }
            }
        });
        __classPrivateFieldGet(this, _RPCConnection_reader, "f").on('error', (error) => {
            __classPrivateFieldGet(this, _RPCConnection_logger, "f").error('[yt-cast-receiver] RPC connection reader error:', error);
            // Force disconnect
            readable.destroy();
            __classPrivateFieldGet(this, _RPCConnection_instances, "m", _RPCConnection_handleDisconnect).call(this);
        });
        __classPrivateFieldGet(this, _RPCConnection_reader, "f").on('end', __classPrivateFieldGet(this, _RPCConnection_instances, "m", _RPCConnection_handleDisconnect).bind(this));
        readable.on('end', __classPrivateFieldGet(this, _RPCConnection_instances, "m", _RPCConnection_handleDisconnect).bind(this));
        return;
    }
    __classPrivateFieldSet(this, _RPCConnection_status, 'disconnected', "f");
    throw new BadResponseError('RPC connection request returned bad response', url, response);
}, _RPCConnection_handleDisconnect = function _RPCConnection_handleDisconnect() {
    if (__classPrivateFieldGet(this, _RPCConnection_status, "f") === 'disconnected') {
        // Already handled
        return;
    }
    const prevStatus = __classPrivateFieldGet(this, _RPCConnection_status, "f");
    __classPrivateFieldSet(this, _RPCConnection_status, 'disconnected', "f");
    if (__classPrivateFieldGet(this, _RPCConnection_reader, "f")) {
        __classPrivateFieldGet(this, _RPCConnection_reader, "f").removeAllListeners();
        __classPrivateFieldSet(this, _RPCConnection_reader, null, "f");
    }
    if (prevStatus === 'connected') {
        // Disconnected by remote end or reader error - reconnect.
        __classPrivateFieldGet(this, _RPCConnection_logger, "f").debug('[yt-cast-receiver] RPC connection disconnected. Reconnecting...');
        void (async () => {
            try {
                await __classPrivateFieldGet(this, _RPCConnection_instances, "m", _RPCConnection_doConnect).call(this, true);
            }
            catch (error) {
                this.emit('terminate', error);
            }
        })();
    }
    else {
        __classPrivateFieldGet(this, _RPCConnection_logger, "f").debug('[yt-cast-receiver] RPC connection closed.');
    }
};
export default RPCConnection;
