// yt-cast-receiver/dist/lib/app/Session.js
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _Session_instances, _Session_client, _Session_screen, _Session_bindParams, _Session_rpcConnection, _Session_ofs, _Session_dataStore, _Session_logger, _Session_taskQueue, _Session_deferredMessages, _Session_loungeTokenRefreshTimer, _Session_pairingCodeRequestService, _Session_status, _Session_refreshLoungeToken, _Session_clearDeferredMessages, _Session_clearLoungeTokenRefreshTimer, _Session_generateScreenId, _Session_getLoungeToken, _Session_getInitSessionMessages, _Session_createSendMessageTask, _Session_doSendMessage, _Session_getSendMessagePayload, _Session_handleMessage, _Session_mdxContext_get, _SendMessageTask_run, _SendMessageTask_cancel, _SendMessageTask_onError;
import EventEmitter from 'events';
import { v4 as uuidv4 } from 'uuid';
import AsyncTaskQueue from '../utils/AsyncTaskQueue.js';
import BindParams from './BindParams.js';
import Message from './Message.js';
import RPCConnection from './RPCConnection.js';
import PairingCodeRequestService from './PairingCodeRequestService.js';
import { BadResponseError, ConnectionError, DataError, IncompleteAPIDataError, SessionError } from '../utils/Errors.js';
import { CLIENTS, STATUSES, URLS } from '../Constants.js';
/**
 * @internal
 *
 * The `Session` class encapsulates the process of setting up the environment for casting:
 *
 * ```
 * const session = new Session(options);
 *
 * // Casting not yet possible
 *
 * await session.begin();
 *
 * // Casting now possible.
 * ```
 *
 * To receive incoming messages, add `messages` event listener:
 * ```
 * session.on('messages', (messages: Message[]) => { // handle messages... });
 * ```
 *
 * To send a message:
 * ```
 * session.sendMessage(message: Message | Message[]);
 * ```
 *
 * To disable casting:
 * ```
 * await session.end();
 * ```
 */
class Session extends EventEmitter {
    constructor(options) {
        super();
        _Session_instances.add(this);
        _Session_client.set(this, void 0);
        _Session_screen.set(this, void 0);
        _Session_bindParams.set(this, void 0);
        _Session_rpcConnection.set(this, void 0);
        _Session_ofs.set(this, void 0);
        _Session_dataStore.set(this, void 0);
        _Session_logger.set(this, void 0);
        _Session_taskQueue.set(this, void 0);
        _Session_deferredMessages.set(this, void 0);
        _Session_loungeTokenRefreshTimer.set(this, void 0);
        _Session_pairingCodeRequestService.set(this, void 0);
        _Session_status.set(this, void 0);
        __classPrivateFieldSet(this, _Session_client, options.client || CLIENTS.YT, "f");
        __classPrivateFieldSet(this, _Session_screen, {
            name: options.screenName,
            app: options.screenApp
        }, "f");
        __classPrivateFieldSet(this, _Session_bindParams, new BindParams({
            theme: __classPrivateFieldGet(this, _Session_client, "f").theme,
            deviceId: uuidv4(),
            screenName: __classPrivateFieldGet(this, _Session_screen, "f").name,
            screenApp: __classPrivateFieldGet(this, _Session_screen, "f").app,
            brand: options.brand,
            model: options.model
        }), "f");
        __classPrivateFieldSet(this, _Session_dataStore, options.dataStore, "f");
        __classPrivateFieldSet(this, _Session_logger, options.logger, "f");
        __classPrivateFieldSet(this, _Session_taskQueue, new AsyncTaskQueue(), "f");
        __classPrivateFieldSet(this, _Session_deferredMessages, {}, "f");
        __classPrivateFieldSet(this, _Session_loungeTokenRefreshTimer, null, "f");
        __classPrivateFieldSet(this, _Session_pairingCodeRequestService, new PairingCodeRequestService(__classPrivateFieldGet(this, _Session_screen, "f"), __classPrivateFieldGet(this, _Session_bindParams, "f")), "f");
        __classPrivateFieldSet(this, _Session_status, STATUSES.STOPPED, "f");
    }
    async begin() {
        const isRefreshing = __classPrivateFieldGet(this, _Session_status, "f") === 'refreshing';
        if (__classPrivateFieldGet(this, _Session_status, "f") !== STATUSES.STOPPED && !isRefreshing) {
            return;
        }
        if (!isRefreshing) {
            __classPrivateFieldSet(this, _Session_status, STATUSES.STARTING, "f");
        }
        // Use stored MDX context if session is not refreshing or restarting (screenId: undefined).
        const mdxContextStorageKey = `mdxContext.${__classPrivateFieldGet(this, _Session_client, "f").theme}`;
        let isScreenIdFromDataStore = false;
        if (!isRefreshing && !__classPrivateFieldGet(this, _Session_screen, "f").id && __classPrivateFieldGet(this, _Session_dataStore, "f")) {
            const storedMdxContext = await __classPrivateFieldGet(this, _Session_dataStore, "f").get(mdxContextStorageKey);
            __classPrivateFieldGet(this, _Session_logger, "f").debug(`[yt-cast-receiver] (${__classPrivateFieldGet(this, _Session_client, "f").name}) Configuring session with stored MDX context:`, storedMdxContext);
            if (storedMdxContext) {
                __classPrivateFieldGet(this, _Session_bindParams, "f").updateWithMdxContext(storedMdxContext);
            }
            if (storedMdxContext?.screenId) {
                __classPrivateFieldGet(this, _Session_screen, "f").id = storedMdxContext.screenId;
                isScreenIdFromDataStore = true;
            }
        }
        // 1. Generate screenId and complete `Screen` info.
        //    - Do not generate if already available, e.g. if refreshing session.
        // 2. With `Screen`, obtain `LoungeToken`.
        // 3. With `Screen` and `LoungeToken`, obtain session-specific params.
        // 4. Finally, with all the above info, create `RPCConnection` to receive remote messages.
        let error;
        try {
            __classPrivateFieldSet(this, _Session_ofs, 0, "f");
            if (!__classPrivateFieldGet(this, _Session_screen, "f").id) {
                __classPrivateFieldGet(this, _Session_screen, "f").id = await __classPrivateFieldGet(this, _Session_instances, "m", _Session_generateScreenId).call(this);
            }
            let loungeToken;
            try {
                loungeToken = await __classPrivateFieldGet(this, _Session_instances, "m", _Session_getLoungeToken).call(this);
            }
            catch (error) {
                if (!isScreenIdFromDataStore) {
                    throw error;
                }
                // Maybe error was caused by using screenId from DataStore.
                // We generate a fresh screenId and try refetching lounge token.
                __classPrivateFieldGet(this, _Session_logger, "f").error(`[yt-cast-receiver] (${__classPrivateFieldGet(this, _Session_client, "f").name}) Failed to obtain lounge token with screen Id from stored MDX context (${__classPrivateFieldGet(this, _Session_screen, "f").id}):`, error, 'Going to generate fresh screen Id and try again...');
                __classPrivateFieldGet(this, _Session_screen, "f").id = await __classPrivateFieldGet(this, _Session_instances, "m", _Session_generateScreenId).call(this);
                loungeToken = await __classPrivateFieldGet(this, _Session_instances, "m", _Session_getLoungeToken).call(this);
            }
            __classPrivateFieldGet(this, _Session_bindParams, "f").updateWithLoungeToken(loungeToken);
            // Refresh `LoungeToken` upon its expiry
            __classPrivateFieldGet(this, _Session_instances, "m", _Session_clearLoungeTokenRefreshTimer).call(this);
            __classPrivateFieldSet(this, _Session_loungeTokenRefreshTimer, setTimeout(() => {
                void (async () => {
                    try {
                        await __classPrivateFieldGet(this, _Session_instances, "m", _Session_refreshLoungeToken).call(this);
                    }
                    catch (error) {
                        // Session must end if `LoungeToken` can't be obtained
                        this.end(error).catch((error) => __classPrivateFieldGet(this, _Session_logger, "f").error('[yt-cast-receiver] Caught error ending session:', error));
                    }
                })();
            }, loungeToken.refreshIntervalInMillis || 1123200000), "f"); // Fallback to 13 days
            const initSessionMessages = await __classPrivateFieldGet(this, _Session_instances, "m", _Session_getInitSessionMessages).call(this);
            for (const cmd of initSessionMessages) {
                if (cmd.name === 'c' || cmd.name === 'S') {
                    __classPrivateFieldGet(this, _Session_bindParams, "f").updateWithMessage(cmd);
                }
                else if (cmd.name === 'loungeStatus') {
                    __classPrivateFieldGet(this, _Session_instances, "m", _Session_handleMessage).call(this, [cmd]);
                }
            }
            // Test if we can start posting state / rpc
            try {
                __classPrivateFieldGet(this, _Session_bindParams, "f").toQueryString('rpc');
            }
            catch (error) {
                if (error instanceof IncompleteAPIDataError) {
                    throw new IncompleteAPIDataError(`(${__classPrivateFieldGet(this, _Session_client, "f").name}) Query string test failed`, (error).info?.missing);
                }
                throw error;
            }
        }
        catch (err) {
            error = new SessionError(`(${__classPrivateFieldGet(this, _Session_client, "f").name}) Failed to establish session`, err);
        }
        if (!error) {
            __classPrivateFieldSet(this, _Session_rpcConnection, new RPCConnection({
                bindParams: __classPrivateFieldGet(this, _Session_bindParams, "f"),
                logger: __classPrivateFieldGet(this, _Session_logger, "f")
            }), "f");
            __classPrivateFieldGet(this, _Session_rpcConnection, "f").on('messages', __classPrivateFieldGet(this, _Session_instances, "m", _Session_handleMessage).bind(this));
            __classPrivateFieldGet(this, _Session_rpcConnection, "f").on('terminate', (error) => {
                __classPrivateFieldGet(this, _Session_logger, "f").error(`[yt-cast-receiver] (${__classPrivateFieldGet(this, _Session_client, "f").name}) RPC connection terminated due to error:`, error);
                __classPrivateFieldGet(this, _Session_instances, "m", _Session_refreshLoungeToken).call(this).catch((error) => __classPrivateFieldGet(this, _Session_logger, "f").error('[yt-cast-receiver] Caught error refreshing lounge token:', error));
            });
            try {
                await __classPrivateFieldGet(this, _Session_rpcConnection, "f").connect();
                __classPrivateFieldGet(this, _Session_logger, "f").debug(`[yt-cast-receiver] (${__classPrivateFieldGet(this, _Session_client, "f").name}) Session established.`);
                if (!isRefreshing) {
                    __classPrivateFieldSet(this, _Session_status, STATUSES.RUNNING, "f");
                }
            }
            catch (err) {
                error = new SessionError(`(${__classPrivateFieldGet(this, _Session_client, "f").name}) Failed to establish session`, err);
            }
        }
        if (error) {
            // Refreshing happens internally, so we need to emit the error instead of
            // Throwing it (i.e. don't pass it to `end()`).
            if (isRefreshing) {
                this.end(error).catch((error) => __classPrivateFieldGet(this, _Session_logger, "f").error('[yt-cast-receiver] Caught error ending session:', error));
                ;
            }
            else {
                try {
                    this.end().catch((error) => __classPrivateFieldGet(this, _Session_logger, "f").error('[yt-cast-receiver] Caught error ending session:', error));
                    ;
                }
                catch (_error) {
                    // Do nothing - we're quitting anyway.
                }
                throw error;
            }
        }
        else if (__classPrivateFieldGet(this, _Session_dataStore, "f")) {
            // Session started successfully - store MDX context.
            const mdxContext = __classPrivateFieldGet(this, _Session_instances, "a", _Session_mdxContext_get);
            __classPrivateFieldGet(this, _Session_logger, "f").debug(`[yt-cast-receiver] (${__classPrivateFieldGet(this, _Session_client, "f").name}) Saving MDX context to data store:`, mdxContext);
            __classPrivateFieldGet(this, _Session_dataStore, "f").set(mdxContextStorageKey, mdxContext)
                .catch((error) => __classPrivateFieldGet(this, _Session_logger, "f").error('[yt-cast-receiver] Caught error saving MDX context:', error));
            ;
        }
    }
    async end(error) {
        if (__classPrivateFieldGet(this, _Session_status, "f") === STATUSES.STOPPED || __classPrivateFieldGet(this, _Session_status, "f") === STATUSES.STOPPING) {
            return;
        }
        __classPrivateFieldSet(this, _Session_status, STATUSES.STOPPING, "f");
        try {
            __classPrivateFieldGet(this, _Session_pairingCodeRequestService, "f").stop();
            if (__classPrivateFieldGet(this, _Session_rpcConnection, "f")) {
                __classPrivateFieldGet(this, _Session_rpcConnection, "f").removeAllListeners();
                __classPrivateFieldGet(this, _Session_rpcConnection, "f").close();
            }
            __classPrivateFieldGet(this, _Session_taskQueue, "f").clear();
            __classPrivateFieldGet(this, _Session_instances, "m", _Session_clearDeferredMessages).call(this);
            try {
                await this.sendMessage(new Message.LoungeScreenDisconnected());
                __classPrivateFieldGet(this, _Session_taskQueue, "f").start() // In case stopped elsewhere (e.g. `#refreshLoungeToken()`)
                    .catch((error) => __classPrivateFieldGet(this, _Session_logger, "f").error('[yt-cast-receiver] Caught error starting task queue:', error));
            }
            catch (_error) {
                // `sendMessage()` should work when status is 'running'.
                // But otherwise there is no guarantee that current state is valid to allow
                // Message to be sent successfully.
            }
        }
        catch (err) {
            if (!error) {
                throw new SessionError(`(${__classPrivateFieldGet(this, _Session_client, "f").name}) Error while ending session`, err);
            }
        }
        finally {
            __classPrivateFieldGet(this, _Session_bindParams, "f").reset();
            __classPrivateFieldSet(this, _Session_status, STATUSES.STOPPED, "f");
        }
        if (error) {
            this.emit('terminate', error);
        }
    }
    async restart() {
        __classPrivateFieldGet(this, _Session_logger, "f").debug(`[yt-cast-receiver] (${__classPrivateFieldGet(this, _Session_client, "f").name}) Restarting session...`);
        const pairingCodeRequestServiceStatus = __classPrivateFieldGet(this, _Session_pairingCodeRequestService, "f").status;
        try {
            await this.end();
            await this.begin();
            if (pairingCodeRequestServiceStatus === STATUSES.RUNNING) {
                __classPrivateFieldGet(this, _Session_pairingCodeRequestService, "f").start();
            }
        }
        catch (error) {
            throw new SessionError(`(${__classPrivateFieldGet(this, _Session_client, "f").name}) Error while restarting session`, error);
        }
    }
    async registerPairingCode(code) {
        __classPrivateFieldGet(this, _Session_logger, "f").debug(`[yt-cast-receiver] (${__classPrivateFieldGet(this, _Session_client, "f").name}) Registering pairing code: ${code}`);
        if (!__classPrivateFieldGet(this, _Session_screen, "f").id) {
            throw new IncompleteAPIDataError(`(${__classPrivateFieldGet(this, _Session_client, "f").name}) Missing data required to register pairing code`, ['screenId']);
        }
        const params = new URLSearchParams({
            'access_type': 'permanent',
            'app': __classPrivateFieldGet(this, _Session_screen, "f").app,
            'pairing_code': code,
            'screen_id': __classPrivateFieldGet(this, _Session_screen, "f").id,
            'screen_name': __classPrivateFieldGet(this, _Session_screen, "f").name,
            'device_id': __classPrivateFieldGet(this, _Session_bindParams, "f").id
        });
        const url = URLS.REGISTER_PAIRING_CODE;
        let response;
        try {
            response = await fetch(url, {
                method: 'POST',
                body: params
            });
        }
        catch (error) {
            throw new ConnectionError(`(${__classPrivateFieldGet(this, _Session_client, "f").name}) Connection error in registering pairing code`, url, error);
        }
        if (response.ok) {
            __classPrivateFieldGet(this, _Session_logger, "f").debug(`[yt-cast-receiver] (${__classPrivateFieldGet(this, _Session_client, "f").name}) Pairing code registered.`);
            return;
        }
        throw new BadResponseError(`(${__classPrivateFieldGet(this, _Session_client, "f").name}) Failed to register pairing code`, url, response);
    }
    /**
     * Sends `messages`. If `defer: {key, interval}` is specified, only send after `defer.interval` (in milliseconds) has lapsed.
     *
     * @param messages - `Message` object or array of `Message` objects.
     * @param defer - Specify `{key, interval}` to defer sending, or `false` to send as soon as possible.
     * @param defer.key - If there is already a deferred message with the same `key`, it will be canceled and replaced with the current one.
     * @param defer.interval - The interval in milliseconds to lapse before sending.
     * @returns A Promise that resolves to `true` when the message has been sent, or `false` if it is cancelled prior to dispatch.
     */
    sendMessage(messages, defer = false) {
        // Use `AsyncTaskQueue` for sending messages. This ensures that messages will be sent in order.
        if (defer) {
            return new Promise((resolve, reject) => {
                const { key, interval } = defer;
                const existing = __classPrivateFieldGet(this, _Session_deferredMessages, "f")[key];
                if (existing) {
                    clearTimeout(existing.timeout);
                    delete __classPrivateFieldGet(this, _Session_deferredMessages, "f")[key];
                    existing.task.cancel();
                }
                const timeout = setTimeout(() => {
                    try {
                        const task = __classPrivateFieldGet(this, _Session_deferredMessages, "f")[key].task;
                        delete __classPrivateFieldGet(this, _Session_deferredMessages, "f")[key];
                        __classPrivateFieldGet(this, _Session_taskQueue, "f").push(task);
                    }
                    catch (error) {
                        reject(error instanceof Error ? error : Error(String(error)));
                    }
                }, interval);
                __classPrivateFieldGet(this, _Session_deferredMessages, "f")[key] = {
                    task: __classPrivateFieldGet(this, _Session_instances, "m", _Session_createSendMessageTask).call(this, messages, resolve, reject),
                    timeout
                };
            });
        }
        return new Promise((resolve, reject) => {
            __classPrivateFieldGet(this, _Session_taskQueue, "f").push(__classPrivateFieldGet(this, _Session_instances, "m", _Session_createSendMessageTask).call(this, messages, resolve, reject));
        });
    }
    get pairingCodeRequestService() {
        return __classPrivateFieldGet(this, _Session_pairingCodeRequestService, "f");
    }
    get client() {
        return __classPrivateFieldGet(this, _Session_client, "f");
    }
    on(event, listener) {
        super.on(event, listener);
        return this;
    }
}
_Session_client = new WeakMap(), _Session_screen = new WeakMap(), _Session_bindParams = new WeakMap(), _Session_rpcConnection = new WeakMap(), _Session_ofs = new WeakMap(), _Session_dataStore = new WeakMap(), _Session_logger = new WeakMap(), _Session_taskQueue = new WeakMap(), _Session_deferredMessages = new WeakMap(), _Session_loungeTokenRefreshTimer = new WeakMap(), _Session_pairingCodeRequestService = new WeakMap(), _Session_status = new WeakMap(), _Session_instances = new WeakSet(), _Session_refreshLoungeToken = async function _Session_refreshLoungeToken() {
    __classPrivateFieldSet(this, _Session_status, 'refreshing', "f");
    __classPrivateFieldGet(this, _Session_logger, "f").debug(`[yt-cast-receiver] (${__classPrivateFieldGet(this, _Session_client, "f").name}) Refreshing lounge token...`);
    __classPrivateFieldGet(this, _Session_instances, "m", _Session_clearLoungeTokenRefreshTimer).call(this);
    const oldRPC = __classPrivateFieldGet(this, _Session_rpcConnection, "f");
    __classPrivateFieldGet(this, _Session_taskQueue, "f").setAutoStart(false);
    __classPrivateFieldGet(this, _Session_taskQueue, "f").stop();
    const pairingCodeRequestServiceStatus = __classPrivateFieldGet(this, _Session_pairingCodeRequestService, "f").status;
    __classPrivateFieldGet(this, _Session_pairingCodeRequestService, "f").stop();
    __classPrivateFieldGet(this, _Session_bindParams, "f").reset();
    try {
        await this.begin();
    }
    catch (error) {
        throw new SessionError(`(${__classPrivateFieldGet(this, _Session_client, "f").name}) Error while refreshing lounge token`, error);
    }
    finally {
        __classPrivateFieldGet(this, _Session_logger, "f").debug(`[yt-cast-receiver] (${__classPrivateFieldGet(this, _Session_client, "f").name}) Closing old RPC connection...`);
        oldRPC.removeAllListeners();
        oldRPC.close();
    }
    if (pairingCodeRequestServiceStatus === STATUSES.RUNNING) {
        __classPrivateFieldGet(this, _Session_pairingCodeRequestService, "f").start();
    }
    // Nullify AIDs of pending Messages, because AID restarts from 0 after refresh.
    Object.values(__classPrivateFieldGet(this, _Session_deferredMessages, "f")).forEach((dc) => {
        const sc = Array.isArray(dc.task.message) ? dc.task.message : [dc.task.message];
        sc.forEach((c) => c.AID = null);
    });
    __classPrivateFieldGet(this, _Session_taskQueue, "f").forEach((task) => {
        if (task instanceof SendMessageTask) {
            const sc = Array.isArray(task.message) ? task.message : [task.message];
            sc.forEach((c) => c.AID = null);
        }
    });
    __classPrivateFieldGet(this, _Session_taskQueue, "f").setAutoStart(true);
    __classPrivateFieldSet(this, _Session_status, STATUSES.RUNNING, "f");
}, _Session_clearDeferredMessages = function _Session_clearDeferredMessages() {
    Object.values(__classPrivateFieldGet(this, _Session_deferredMessages, "f")).forEach((c) => {
        clearTimeout(c.timeout);
        c.task.cancel();
    });
    __classPrivateFieldSet(this, _Session_deferredMessages, {}, "f");
}, _Session_clearLoungeTokenRefreshTimer = function _Session_clearLoungeTokenRefreshTimer() {
    if (__classPrivateFieldGet(this, _Session_loungeTokenRefreshTimer, "f")) {
        clearTimeout(__classPrivateFieldGet(this, _Session_loungeTokenRefreshTimer, "f"));
        __classPrivateFieldSet(this, _Session_loungeTokenRefreshTimer, null, "f");
    }
}, _Session_generateScreenId = async function _Session_generateScreenId() {
    const url = URLS.GENERATE_SCREEN_ID;
    let response;
    try {
        response = await fetch(url);
    }
    catch (error) {
        throw new ConnectionError(`(${__classPrivateFieldGet(this, _Session_client, "f").name}) Connection error in generating screen Id`, url, error);
    }
    try {
        const screenId = await response.text();
        __classPrivateFieldGet(this, _Session_logger, "f").debug(`[yt-cast-receiver] (${__classPrivateFieldGet(this, _Session_client, "f").name}) Obtained screen ID: ${screenId}`);
        return screenId;
    }
    catch (error) {
        throw new DataError(`(${__classPrivateFieldGet(this, _Session_client, "f").name}) Screen Id data error`, error);
    }
}, _Session_getLoungeToken = async function _Session_getLoungeToken() {
    if (!__classPrivateFieldGet(this, _Session_screen, "f").id) {
        throw new IncompleteAPIDataError(`(${__classPrivateFieldGet(this, _Session_client, "f").name}) Missing data required to get lounge token`, ['screenId']);
    }
    const url = URLS.GET_LOUNGE_TOKEN_BATCH;
    const params = new URLSearchParams({
        'screen_ids': __classPrivateFieldGet(this, _Session_screen, "f").id
    });
    let response;
    try {
        response = await fetch(URLS.GET_LOUNGE_TOKEN_BATCH, {
            method: 'POST',
            body: params
        });
    }
    catch (error) {
        throw new ConnectionError(`(${__classPrivateFieldGet(this, _Session_client, "f").name}) Connection error in getting lounge token`, url, error);
    }
    try {
        const token = (await response.json())?.screens?.[0];
        __classPrivateFieldGet(this, _Session_logger, "f").debug(`[yt-cast-receiver] (${__classPrivateFieldGet(this, _Session_client, "f").name}) Obtained lounge token:`, token);
        return token;
    }
    catch (error) {
        throw new DataError(`(${__classPrivateFieldGet(this, _Session_client, "f").name}) Lounge token data error`, error);
    }
}, _Session_getInitSessionMessages = async function _Session_getInitSessionMessages() {
    const url = `${URLS.BIND}?${__classPrivateFieldGet(this, _Session_bindParams, "f").toQueryString('initSession')}`;
    let response;
    try {
        response = await fetch(url, {
            method: 'POST',
            body: new URLSearchParams({ count: '0' })
        });
    }
    catch (error) {
        throw new ConnectionError(`(${__classPrivateFieldGet(this, _Session_client, "f").name}) Connection error in fetching session data`, url, error);
    }
    try {
        const body = await response.text();
        const messages = Message.parseIncoming(body);
        __classPrivateFieldGet(this, _Session_logger, "f").debug(`[yt-cast-receiver] (${__classPrivateFieldGet(this, _Session_client, "f").name}) Received messages for establishing session:`, messages);
        return messages;
    }
    catch (error) {
        throw new DataError(`(${__classPrivateFieldGet(this, _Session_client, "f").name}) Session data error`, error);
    }
}, _Session_createSendMessageTask = function _Session_createSendMessageTask(sc, resolve, reject) {
    return new SendMessageTask({
        message: sc,
        run: __classPrivateFieldGet(this, _Session_instances, "m", _Session_doSendMessage).bind(this, sc, resolve, reject),
        cancel: () => {
            resolve(false);
        },
        onError: (task, error) => {
            const smTask = task;
            __classPrivateFieldGet(this, _Session_logger, "f").error(`[yt-cast-receiver] (${this.client.name}) Error occurred in SendMessageTask:`, smTask.message, error);
            // Retry task after refreshToken, but if it fails again then we would have to end session.
            smTask.onError = (task, error) => {
                this.end(error).catch((error) => __classPrivateFieldGet(this, _Session_logger, "f").error('[yt-cast-receiver] Caught error ending session:', error));
            };
            __classPrivateFieldGet(this, _Session_taskQueue, "f").unshift(smTask);
            __classPrivateFieldGet(this, _Session_logger, "f").debug(`[yt-cast-receiver] (${__classPrivateFieldGet(this, _Session_client, "f").name}) Retry task after refreshing lounge token...`);
            try {
                __classPrivateFieldGet(this, _Session_instances, "m", _Session_refreshLoungeToken).call(this).catch((error) => __classPrivateFieldGet(this, _Session_logger, "f").error('[yt-cast-receiver] Caught error refreshing lounge token:', error));
            }
            catch (err) {
                this.end(err).catch((error) => __classPrivateFieldGet(this, _Session_logger, "f").error('[yt-cast-receiver] Caught error ending session:', error));
            }
        }
    });
}, _Session_doSendMessage = async function _Session_doSendMessage(sc, resolve, reject) {
    const AID = Array.isArray(sc) ? sc[0]?.AID : sc.AID;
    const url = `${URLS.BIND}?${__classPrivateFieldGet(this, _Session_bindParams, "f").toQueryString('sendMessage', AID)}`;
    const postData = __classPrivateFieldGet(this, _Session_instances, "m", _Session_getSendMessagePayload).call(this, sc);
    const debugMsgNameStr = Array.isArray(sc) ? `messages '${sc.map((c) => c.name).join(' + ')}'` : `message '${sc.name}'`;
    __classPrivateFieldGet(this, _Session_logger, "f").debug(`[yt-cast-receiver] ${AID ? `(AID: ${AID}) ` : ''}(${__classPrivateFieldGet(this, _Session_client, "f").name}) Sending ${debugMsgNameStr} with payload:`, postData);
    let response;
    try {
        response = await fetch(url, {
            method: 'POST',
            body: new URLSearchParams(postData)
        });
        __classPrivateFieldGet(this, _Session_logger, "f").debug(`[yt-cast-receiver] ${AID ? `(AID: ${AID}) ` : ''}(${__classPrivateFieldGet(this, _Session_client, "f").name}) Response received for sent ${debugMsgNameStr}. Status: ${response.status}`);
    }
    catch (error) {
        reject(new ConnectionError(`(${__classPrivateFieldGet(this, _Session_client, "f").name}) Connection error in sending ${debugMsgNameStr}${AID ? ` (AID: ${AID})` : ''}`, url, error));
        return;
    }
    if (response.ok) {
        resolve(true);
    }
    reject(new BadResponseError(`(${__classPrivateFieldGet(this, _Session_client, "f").name}) Bad response received for ${debugMsgNameStr}${AID ? ` (AID: ${AID})` : ''}`, url, response));
}, _Session_getSendMessagePayload = function _Session_getSendMessagePayload(sc) {
    const states = Array.isArray(sc) ? sc : [sc];
    const postData = {
        'count': states.length,
        'ofs': `${__classPrivateFieldGet(this, _Session_ofs, "f")}`
    };
    states.forEach((state, index) => {
        const keyPrefix = `req${index}_`;
        postData[`${keyPrefix}_sc`] = state.name;
        for (const [key, value] of Object.entries(state.payload)) {
            postData[keyPrefix + key] = value;
        }
    });
    __classPrivateFieldSet(this, _Session_ofs, __classPrivateFieldGet(this, _Session_ofs, "f") + states.length, "f");
    return postData;
}, _Session_handleMessage = function _Session_handleMessage(messages) {
    __classPrivateFieldGet(this, _Session_bindParams, "f").updateWithMessage(messages);
    this.emit('messages', messages, this);
}, _Session_mdxContext_get = function _Session_mdxContext_get() {
    return {
        deviceId: __classPrivateFieldGet(this, _Session_bindParams, "f").id,
        screenId: __classPrivateFieldGet(this, _Session_screen, "f").id
    };
};
export default Session;
/**
 * @internal
 *
 * An {@link AsyncTaskQueue} {@link Task} that sends a {@link Message}.
 */
export class SendMessageTask {
    constructor(data) {
        _SendMessageTask_run.set(this, void 0);
        _SendMessageTask_cancel.set(this, void 0);
        _SendMessageTask_onError.set(this, void 0);
        this.message = data.message;
        __classPrivateFieldSet(this, _SendMessageTask_run, data.run, "f");
        __classPrivateFieldSet(this, _SendMessageTask_cancel, data.cancel, "f");
        __classPrivateFieldSet(this, _SendMessageTask_onError, data.onError, "f");
    }
    async run() {
        await __classPrivateFieldGet(this, _SendMessageTask_run, "f").call(this);
    }
    cancel() {
        __classPrivateFieldGet(this, _SendMessageTask_cancel, "f").call(this);
    }
    onError(task, error) {
        __classPrivateFieldGet(this, _SendMessageTask_onError, "f").call(this, task, error);
    }
}
_SendMessageTask_run = new WeakMap(), _SendMessageTask_cancel = new WeakMap(), _SendMessageTask_onError = new WeakMap();
