// yt-cast-receiver/dist/lib/app/YouTubeApp.js
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _YouTubeApp_instances, _YouTubeApp_sessions, _YouTubeApp_activeSession, _YouTubeApp_connectedSenders, _YouTubeApp_dataStore, _YouTubeApp_logger, _YouTubeApp_player, _YouTubeApp_autoplayModeOnConnect, _YouTubeApp_mutePolicy, _YouTubeApp_resetPlayerOnDisconnectPolicy, _YouTubeApp_implicitlyDisconnectedSenders, _YouTubeApp_autoplayModeBeforeUnsupportedOverride, _YouTubeApp_playerStateListener, _YouTubeApp_checkAndSwitchActiveSession, _YouTubeApp_setAutoplayMode, _YouTubeApp_setAutoplayModeBySenderCapabilities, _YouTubeApp_enforceMutePolicy, _YouTubeApp_handleIncomingMessage, _YouTubeApp_handleSenderConnected, _YouTubeApp_handleSenderDisconnected, _YouTubeApp_handlePlayerStateEvent;
import EventEmitter from 'events';
import queryString from 'query-string';
import { v4 as uuidv4 } from 'uuid';
import Message from './Message.js';
import Session from './Session.js';
import Sender from './Sender.js';
import { AppError, SenderConnectionError, IncompleteAPIDataError } from '../utils/Errors.js';
import { AUTOPLAY_MODES, STATUSES, CONF_DEFAULTS, PLAYER_STATUSES, CLIENTS, MUTE_POLICIES, RESET_PLAYER_ON_DISCONNECT_POLICIES } from '../Constants.js';
import DefaultPlaylistRequestHandler from './DefaultPlaylistRequestHandler.js';
/**
 * @internal
 *
 * A `YouTubeApp` instance readies the cast environment, handles incoming messages
 * (responding to them as necessary), posts status updates, manages player state, etc..
 */
class YouTubeApp extends EventEmitter {
    constructor(player, options) {
        super();
        _YouTubeApp_instances.add(this);
        _YouTubeApp_sessions.set(this, void 0);
        _YouTubeApp_activeSession.set(this, void 0);
        _YouTubeApp_connectedSenders.set(this, void 0);
        _YouTubeApp_dataStore.set(this, void 0);
        _YouTubeApp_logger.set(this, void 0);
        _YouTubeApp_player.set(this, void 0);
        _YouTubeApp_autoplayModeOnConnect.set(this, void 0);
        _YouTubeApp_mutePolicy.set(this, void 0);
        _YouTubeApp_resetPlayerOnDisconnectPolicy.set(this, void 0);
        _YouTubeApp_implicitlyDisconnectedSenders.set(this, void 0);
        _YouTubeApp_autoplayModeBeforeUnsupportedOverride.set(this, void 0);
        _YouTubeApp_playerStateListener.set(this, void 0);
        this.name = 'YouTube Cast Receiver App';
        this.state = STATUSES.STOPPED;
        this.allowStop = false;
        __classPrivateFieldSet(this, _YouTubeApp_connectedSenders, [], "f");
        __classPrivateFieldSet(this, _YouTubeApp_implicitlyDisconnectedSenders, [], "f");
        __classPrivateFieldSet(this, _YouTubeApp_dataStore, options.dataStore || null, "f");
        __classPrivateFieldSet(this, _YouTubeApp_logger, options.logger, "f");
        const commonSessionOptions = {
            screenName: options.screenName,
            screenApp: options.screenApp || CONF_DEFAULTS.SCREEN_APP,
            brand: options.brand || CONF_DEFAULTS.BRAND,
            model: options.model || CONF_DEFAULTS.MODEL,
            dataStore: __classPrivateFieldGet(this, _YouTubeApp_dataStore, "f"),
            logger: __classPrivateFieldGet(this, _YouTubeApp_logger, "f")
        };
        __classPrivateFieldSet(this, _YouTubeApp_sessions, {
            YT: new Session({ client: CLIENTS.YT, ...commonSessionOptions }),
            YTMUSIC: new Session({ client: CLIENTS.YTMUSIC, ...commonSessionOptions })
        }, "f");
        __classPrivateFieldSet(this, _YouTubeApp_activeSession, null, "f");
        __classPrivateFieldSet(this, _YouTubeApp_player, player, "f");
        __classPrivateFieldGet(this, _YouTubeApp_player, "f").setLogger(options.logger);
        __classPrivateFieldGet(this, _YouTubeApp_player, "f").queue.setLogger(options.logger);
        __classPrivateFieldGet(this, _YouTubeApp_player, "f").queue.setRequestHandler(options.playlistRequestHandler || new DefaultPlaylistRequestHandler());
        __classPrivateFieldGet(this, _YouTubeApp_player, "f").queue.requestHandler.setLogger(options.logger);
        __classPrivateFieldSet(this, _YouTubeApp_playerStateListener, __classPrivateFieldGet(this, _YouTubeApp_instances, "m", _YouTubeApp_handlePlayerStateEvent).bind(this), "f");
        this.enableAutoplayOnConnect(options.enableAutoplayOnConnect !== undefined ? options.enableAutoplayOnConnect : true);
        __classPrivateFieldSet(this, _YouTubeApp_mutePolicy, options.mutePolicy !== undefined ? options.mutePolicy : MUTE_POLICIES.AUTO, "f");
        __classPrivateFieldSet(this, _YouTubeApp_autoplayModeBeforeUnsupportedOverride, null, "f");
        this.setResetPlayerOnDisconnectPolicy(options.resetPlayerOnDisconnectPolicy);
    }
    async start() {
        if (this.state !== STATUSES.STOPPED) {
            return;
        }
        if (__classPrivateFieldGet(this, _YouTubeApp_dataStore, "f")) {
            const storedPid = await __classPrivateFieldGet(this, _YouTubeApp_dataStore, "f").get('app.pid');
            if (storedPid) {
                __classPrivateFieldGet(this, _YouTubeApp_logger, "f").debug(`[yt-cast-receiver] Setting app pid to stored value: ${storedPid}`);
                this.pid = storedPid;
            }
            else {
                this.pid = uuidv4();
                __classPrivateFieldGet(this, _YouTubeApp_logger, "f").debug(`[yt-cast-receiver] Saving generated app pid: ${this.pid}`);
                await __classPrivateFieldGet(this, _YouTubeApp_dataStore, "f").set('app.pid', this.pid);
            }
        }
        __classPrivateFieldGet(this, _YouTubeApp_logger, "f").debug('[yt-cast-receiver] Starting YouTubeApp...');
        this.state = STATUSES.STARTING;
        __classPrivateFieldGet(this, _YouTubeApp_player, "f").on('state', __classPrivateFieldGet(this, _YouTubeApp_playerStateListener, "f"));
        const clientKeys = Object.keys(CLIENTS);
        try {
            const startPromises = clientKeys.map((key) => {
                const session = __classPrivateFieldGet(this, _YouTubeApp_sessions, "f")[key];
                session.on('messages', (message, session) => {
                    __classPrivateFieldGet(this, _YouTubeApp_instances, "m", _YouTubeApp_handleIncomingMessage).call(this, message, session)
                        .catch((error) => __classPrivateFieldGet(this, _YouTubeApp_logger, "f").error('[yt-cast-receiver] Caught error handling incoming message:', error));
                });
                session.on('terminate', (error) => {
                    // Session terminated. So should YouTubeApp.
                    this.stop(error)
                        .catch((error) => __classPrivateFieldGet(this, _YouTubeApp_logger, "f").error('[yt-cast-receiver] Caught error terminating YouTubeApp:', error));
                });
                return session.begin();
            });
            await Promise.all(startPromises);
        }
        catch (error) {
            __classPrivateFieldGet(this, _YouTubeApp_player, "f").removeListener('state', __classPrivateFieldGet(this, _YouTubeApp_playerStateListener, "f"));
            clientKeys.forEach((key) => {
                void (async () => {
                    const session = __classPrivateFieldGet(this, _YouTubeApp_sessions, "f")[key];
                    session.removeAllListeners();
                    try {
                        await session.end();
                    }
                    catch (_error) {
                        // Do nothing
                    }
                })();
            });
            this.state = STATUSES.STOPPED;
            throw new AppError('Failed to start YouTubeApp', error);
        }
        this.state = STATUSES.RUNNING;
    }
    // Implements
    // eslint-disable-next-line @typescript-eslint/no-misused-promises
    async launch(data) {
        const launchData = queryString.parse(data);
        __classPrivateFieldGet(this, _YouTubeApp_logger, "f").debug('[yt-cast-receiver] YouTubeApp received DIAL launch request. Launch data:', launchData);
        const { pairingCode: pairingCodeData, theme: themeData } = launchData;
        const pairingCode = Array.isArray(pairingCodeData) ? pairingCodeData[0] : pairingCodeData;
        const theme = (Array.isArray(themeData) ? themeData[0] : themeData);
        const session = Object.values(__classPrivateFieldGet(this, _YouTubeApp_sessions, "f")).find((s) => s.client.theme === theme);
        if (pairingCode && session) {
            await __classPrivateFieldGet(this, _YouTubeApp_instances, "m", _YouTubeApp_checkAndSwitchActiveSession).call(this, session);
            __classPrivateFieldGet(this, _YouTubeApp_logger, "f").info('[yt-cast-receiver] Connecting sender through DIAL...');
            await __classPrivateFieldGet(this, _YouTubeApp_activeSession, "f")?.registerPairingCode(pairingCode);
            return this.pid;
        }
        if (!pairingCode) {
            throw new AppError('Failed to launch YouTubeApp', new IncompleteAPIDataError('Invalid launch data', ['pairingCode']));
        }
        throw new AppError('Failed to launch YouTubeApp', new IncompleteAPIDataError(`Invalid launch data. Unknown theme: ${theme}`));
    }
    enableAutoplayOnConnect(value = false) {
        __classPrivateFieldSet(this, _YouTubeApp_autoplayModeOnConnect, value ? AUTOPLAY_MODES.ENABLED : AUTOPLAY_MODES.DISABLED, "f");
    }
    setResetPlayerOnDisconnectPolicy(value) {
        __classPrivateFieldSet(this, _YouTubeApp_resetPlayerOnDisconnectPolicy, value !== undefined ? value : RESET_PLAYER_ON_DISCONNECT_POLICIES.ALL_DISCONNECTED, "f");
    }
    async stop(error) {
        if (this.state !== STATUSES.RUNNING && !error) {
            return;
        }
        __classPrivateFieldGet(this, _YouTubeApp_logger, "f").debug('[yt-cast-receiver] Stopping YouTubeApp...');
        this.state = STATUSES.STOPPING;
        const senders = [...__classPrivateFieldGet(this, _YouTubeApp_connectedSenders, "f")];
        __classPrivateFieldSet(this, _YouTubeApp_connectedSenders, [], "f");
        __classPrivateFieldSet(this, _YouTubeApp_implicitlyDisconnectedSenders, [], "f");
        __classPrivateFieldGet(this, _YouTubeApp_player, "f").removeListener('state', __classPrivateFieldGet(this, _YouTubeApp_playerStateListener, "f"));
        await __classPrivateFieldGet(this, _YouTubeApp_player, "f").reset();
        const stopPromises = Object.values(__classPrivateFieldGet(this, _YouTubeApp_sessions, "f")).map((session) => {
            session.removeAllListeners();
            return session.end();
        });
        try {
            await Promise.all(stopPromises);
        }
        catch (_error) {
            __classPrivateFieldGet(this, _YouTubeApp_logger, "f").warn('[yt-cast-receiver] Ignoring error while stopping YouTubeApp:', error);
        }
        this.state = STATUSES.STOPPED;
        senders.forEach((c) => {
            this.emit('senderDisconnect', c, false);
        });
        if (error) {
            this.emit('terminate', error);
        }
    }
    getPairingCodeRequestService() {
        return __classPrivateFieldGet(this, _YouTubeApp_sessions, "f")['YT'].pairingCodeRequestService;
    }
    get connectedSenders() {
        return [...__classPrivateFieldGet(this, _YouTubeApp_connectedSenders, "f")];
    }
    emit(event, ...args) {
        return super.emit(event, ...args);
    }
    on(event, listener) {
        super.on(event, listener);
        return this;
    }
}
_YouTubeApp_sessions = new WeakMap(), _YouTubeApp_activeSession = new WeakMap(), _YouTubeApp_connectedSenders = new WeakMap(), _YouTubeApp_dataStore = new WeakMap(), _YouTubeApp_logger = new WeakMap(), _YouTubeApp_player = new WeakMap(), _YouTubeApp_autoplayModeOnConnect = new WeakMap(), _YouTubeApp_mutePolicy = new WeakMap(), _YouTubeApp_resetPlayerOnDisconnectPolicy = new WeakMap(), _YouTubeApp_implicitlyDisconnectedSenders = new WeakMap(), _YouTubeApp_autoplayModeBeforeUnsupportedOverride = new WeakMap(), _YouTubeApp_playerStateListener = new WeakMap(), _YouTubeApp_instances = new WeakSet(), _YouTubeApp_checkAndSwitchActiveSession = async function _YouTubeApp_checkAndSwitchActiveSession(targetSession) {
    if (__classPrivateFieldGet(this, _YouTubeApp_activeSession, "f")?.client === targetSession.client) {
        return;
    }
    if (!__classPrivateFieldGet(this, _YouTubeApp_activeSession, "f")) {
        __classPrivateFieldSet(this, _YouTubeApp_activeSession, targetSession, "f");
    }
    // Restart active session (thereby disconnecting current senders) if it is different from target session.
    // We cannot have multiple connections from different clients.
    else {
        __classPrivateFieldGet(this, _YouTubeApp_logger, "f").debug(`[yt-cast-receiver] Target session is for ${targetSession.client.name} clients, whereas active session is for ${__classPrivateFieldGet(this, _YouTubeApp_activeSession, "f").client.name} clients.`);
        __classPrivateFieldGet(this, _YouTubeApp_logger, "f").debug('[yt-cast-receiver] Switching over to target session, while disconnecting senders (if any) from old session.');
        await __classPrivateFieldGet(this, _YouTubeApp_player, "f").reset();
        __classPrivateFieldSet(this, _YouTubeApp_connectedSenders, [], "f");
        __classPrivateFieldSet(this, _YouTubeApp_implicitlyDisconnectedSenders, [], "f");
        const oldSession = __classPrivateFieldGet(this, _YouTubeApp_activeSession, "f");
        __classPrivateFieldGet(this, _YouTubeApp_connectedSenders, "f").forEach((c) => {
            this.emit('senderDisconnect', c, false);
        });
        __classPrivateFieldSet(this, _YouTubeApp_activeSession, targetSession, "f");
        oldSession.sendMessage(new Message.LoungeScreenDisconnected())
            .catch((error) => __classPrivateFieldGet(this, _YouTubeApp_logger, "f").error('[yt-cast-receiver] Caught error sending message:', error));
    }
    __classPrivateFieldGet(this, _YouTubeApp_logger, "f").debug(`[yt-cast-receiver] Active session switched to '${targetSession.client.name}'.`);
}, _YouTubeApp_setAutoplayMode = async function _YouTubeApp_setAutoplayMode(AID, value) {
    if (!__classPrivateFieldGet(this, _YouTubeApp_activeSession, "f")) {
        return;
    }
    const stateBefore = __classPrivateFieldGet(this, _YouTubeApp_player, "f").queue.getState();
    await __classPrivateFieldGet(this, _YouTubeApp_player, "f").queue.setAutoplayMode(value);
    const stateAfter = __classPrivateFieldGet(this, _YouTubeApp_player, "f").queue.getState();
    const sendMessages = [
        new Message.OnAutoplayModeChanged(AID, value)
    ];
    if (stateBefore.autoplay?.id !== stateAfter.autoplay?.id) {
        sendMessages.push(new Message.AutoplayUpNext(AID, stateAfter.autoplay?.id || null));
    }
    __classPrivateFieldGet(this, _YouTubeApp_activeSession, "f").sendMessage(sendMessages)
        .catch((error) => __classPrivateFieldGet(this, _YouTubeApp_logger, "f").error('[yt-cast-receiver] Caught error sending message:', error));
    ;
}, _YouTubeApp_setAutoplayModeBySenderCapabilities = async function _YouTubeApp_setAutoplayModeBySenderCapabilities(AID) {
    if (__classPrivateFieldGet(this, _YouTubeApp_connectedSenders, "f").length === 0) {
        return;
    }
    __classPrivateFieldGet(this, _YouTubeApp_logger, "f").debug('[yt-cast-receiver] Setting autoplay mode by sender(s) capabilities.');
    let autoplayMode;
    if (!__classPrivateFieldGet(this, _YouTubeApp_connectedSenders, "f").every((c) => c.supportsAutoplay())) {
        __classPrivateFieldGet(this, _YouTubeApp_logger, "f").debug('[yt-cast-receiver] (Some) sender(s) do not support autoplay. Autoplay support disabled.');
        if (__classPrivateFieldGet(this, _YouTubeApp_player, "f").autoplayMode !== AUTOPLAY_MODES.UNSUPPORTED) {
            __classPrivateFieldSet(this, _YouTubeApp_autoplayModeBeforeUnsupportedOverride, __classPrivateFieldGet(this, _YouTubeApp_player, "f").autoplayMode, "f");
        }
        autoplayMode = AUTOPLAY_MODES.UNSUPPORTED;
    }
    else {
        __classPrivateFieldGet(this, _YouTubeApp_logger, "f").debug('[yt-cast-receiver] (All) sender(s) support autoplay.');
        if (__classPrivateFieldGet(this, _YouTubeApp_player, "f").autoplayMode !== AUTOPLAY_MODES.UNSUPPORTED) {
            __classPrivateFieldGet(this, _YouTubeApp_logger, "f").debug(`[yt-cast-receiver] Keeping current autoplay mode: ${__classPrivateFieldGet(this, _YouTubeApp_player, "f").autoplayMode}`);
            autoplayMode = __classPrivateFieldGet(this, _YouTubeApp_player, "f").autoplayMode;
        }
        else {
            autoplayMode = __classPrivateFieldGet(this, _YouTubeApp_autoplayModeBeforeUnsupportedOverride, "f") || __classPrivateFieldGet(this, _YouTubeApp_autoplayModeOnConnect, "f");
            __classPrivateFieldGet(this, _YouTubeApp_logger, "f").debug(`[yt-cast-receiver] Setting autoplay mode: ${autoplayMode}`);
        }
        __classPrivateFieldSet(this, _YouTubeApp_autoplayModeBeforeUnsupportedOverride, null, "f");
    }
    return __classPrivateFieldGet(this, _YouTubeApp_instances, "m", _YouTubeApp_setAutoplayMode).call(this, AID, autoplayMode);
}, _YouTubeApp_enforceMutePolicy = async function _YouTubeApp_enforceMutePolicy(AID) {
    if (__classPrivateFieldGet(this, _YouTubeApp_connectedSenders, "f").length === 0) {
        return;
    }
    __classPrivateFieldGet(this, _YouTubeApp_logger, "f").debug('[yt-cast-receiver] Configuring player based on mute policy...');
    const logTrue = 'player will set volume level to 0 on mute';
    const logFalse = 'player will preserve volume level on mute';
    let zeroVolumeLevelOnMute;
    switch (__classPrivateFieldGet(this, _YouTubeApp_mutePolicy, "f")) {
        case MUTE_POLICIES.ZERO_VOLUME_LEVEL:
            __classPrivateFieldGet(this, _YouTubeApp_logger, "f").debug(`[yt-cast-receiver] Mute policy is 'ZERO_VOLUME_LEVEL': ${logTrue}.`);
            zeroVolumeLevelOnMute = true;
            break;
        case MUTE_POLICIES.PRESERVE_VOLUME_LEVEL:
            __classPrivateFieldGet(this, _YouTubeApp_logger, "f").debug(`[yt-cast-receiver] Mute policy is 'PRESERVE_VOLUME_LEVEL': ${logFalse}.`);
            zeroVolumeLevelOnMute = false;
            break;
        default: // MUTE_POLICIES.AUTO
            __classPrivateFieldGet(this, _YouTubeApp_logger, "f").debug('[yt-cast-receiver] Mute policy is \'AUTO\': checking whether sender(s) support mute...');
            if (!__classPrivateFieldGet(this, _YouTubeApp_connectedSenders, "f").every((c) => c.supportsMute())) {
                __classPrivateFieldGet(this, _YouTubeApp_logger, "f").debug(`[yt-cast-receiver] (Some) sender(s) do not support mute: ${logTrue}.`);
                zeroVolumeLevelOnMute = true;
            }
            else {
                __classPrivateFieldGet(this, _YouTubeApp_logger, "f").debug(`[yt-cast-receiver] (All) sender(s) support mute: ${logFalse}.`);
                zeroVolumeLevelOnMute = false;
            }
    }
    await __classPrivateFieldGet(this, _YouTubeApp_player, "f").setZeroVolumeLevelOnMute(zeroVolumeLevelOnMute, AID);
}, _YouTubeApp_handleIncomingMessage = async function _YouTubeApp_handleIncomingMessage(message, session) {
    if (Array.isArray(message)) {
        for (const c of message) {
            await __classPrivateFieldGet(this, _YouTubeApp_instances, "m", _YouTubeApp_handleIncomingMessage).call(this, c, session);
        }
        return;
    }
    const { AID, name, payload } = message;
    const isSessionActive = session === __classPrivateFieldGet(this, _YouTubeApp_activeSession, "f");
    const client = session.client;
    __classPrivateFieldGet(this, _YouTubeApp_logger, "f").debug('-----------------------------------');
    __classPrivateFieldGet(this, _YouTubeApp_logger, "f").debug(`[yt-cast-receiver] (AID: ${AID}) (${client.name}) Incoming message: '${name}'`);
    const sendMessages = [];
    switch (name) {
        case 'getNowPlaying': {
            sendMessages.push(new Message.NowPlaying(AID, isSessionActive ? await __classPrivateFieldGet(this, _YouTubeApp_player, "f").getState() : null));
            break;
        }
        case 'loungeStatus': {
            let loungeDevices;
            try {
                loungeDevices = JSON.parse(payload.devices);
            }
            catch (error) {
                __classPrivateFieldGet(this, _YouTubeApp_logger, "f").error(`[yt-cast-receiver] (${client.name}) Failed to parse 'devices' property of 'loungeStatus' message payload:`, error);
                loungeDevices = [];
            }
            const loungeStatusSenders = loungeDevices
                .filter((dev) => dev.type === 'REMOTE_CONTROL')
                .reduce((senders, data) => {
                try {
                    const sender = Sender.parse(data);
                    senders.push(sender);
                    return senders;
                }
                catch (err) {
                    __classPrivateFieldGet(this, _YouTubeApp_logger, "f").error(`[yt-cast-receiver] (${client.name}) Failed to parse sender data in 'loungeStatus' message:`, err);
                    return senders;
                }
            }, []);
            if (this.state === STATUSES.STARTING) {
                if (loungeStatusSenders.length > 0) {
                    __classPrivateFieldSet(this, _YouTubeApp_connectedSenders, loungeStatusSenders, "f");
                    __classPrivateFieldGet(this, _YouTubeApp_logger, "f").debug(`[yt-cast-receiver] (${client.name}) Updated connected senders info with 'loungeStatus' message:`, __classPrivateFieldGet(this, _YouTubeApp_connectedSenders, "f"));
                    __classPrivateFieldSet(this, _YouTubeApp_activeSession, session, "f");
                    __classPrivateFieldGet(this, _YouTubeApp_logger, "f").debug(`[yt-cast-receiver] Active session switched to '${client.name}'.`);
                    await __classPrivateFieldGet(this, _YouTubeApp_instances, "m", _YouTubeApp_setAutoplayModeBySenderCapabilities).call(this, AID);
                    await __classPrivateFieldGet(this, _YouTubeApp_instances, "m", _YouTubeApp_enforceMutePolicy).call(this, AID);
                    sendMessages.push(new Message.OnHasPreviousNextChanged(AID, __classPrivateFieldGet(this, _YouTubeApp_player, "f").getNavInfo()), new Message.NowPlaying(AID, null));
                }
            }
            else {
                let connectionEventDetails;
                try {
                    connectionEventDetails = payload.connectionEventDetails ? JSON.parse(payload.connectionEventDetails) : null;
                }
                catch (err) {
                    const error = new SenderConnectionError('Failed to handle sender connection / disconnection', err);
                    __classPrivateFieldGet(this, _YouTubeApp_logger, "f").error('[yt-cast-receiver] Failed to parse connection event details in payload of loungeStatus message:', payload.connectionEventDetails);
                    this.emit('error', error);
                    connectionEventDetails = null;
                }
                let handledConnectionEvent = false;
                if (connectionEventDetails?.deviceId) {
                    const connectingSender = loungeStatusSenders.find((s) => s.id === connectionEventDetails.deviceId);
                    const disconnectingSender = !connectingSender ? __classPrivateFieldGet(this, _YouTubeApp_connectedSenders, "f").find((s) => s.id === connectionEventDetails.deviceId) : null;
                    if (connectingSender) {
                        sendMessages.push(...await __classPrivateFieldGet(this, _YouTubeApp_instances, "m", _YouTubeApp_handleSenderConnected).call(this, connectingSender, session, AID));
                        handledConnectionEvent = true;
                    }
                    if (disconnectingSender) {
                        const isImplicitDisconnect = connectionEventDetails.ui !== 'true';
                        sendMessages.push(...await __classPrivateFieldGet(this, _YouTubeApp_instances, "m", _YouTubeApp_handleSenderDisconnected).call(this, disconnectingSender, session, isImplicitDisconnect, AID));
                        handledConnectionEvent = true;
                    }
                }
                if (!handledConnectionEvent) {
                    const playerNavInfo = isSessionActive ? __classPrivateFieldGet(this, _YouTubeApp_player, "f").getNavInfo() : null;
                    sendMessages.push(new Message.OnHasPreviousNextChanged(AID, playerNavInfo), new Message.OnAutoplayModeChanged(AID, playerNavInfo));
                }
            }
            break;
        }
        case 'setPlaylist':
        case 'updatePlaylist': {
            if (!isSessionActive)
                return;
            __classPrivateFieldGet(this, _YouTubeApp_logger, "f").debug(`[yt-cast-receiver] '${message.name}' message payload:`, payload);
            // Dismiss autoplay first in anticipation of change in autoplay video (when last video in queue
            // Is different from the one in message payload).
            const queueVideoIds = __classPrivateFieldGet(this, _YouTubeApp_player, "f").queue.videoIds;
            const msgPayloadLastVideoId = message.payload?.videoIds?.split(',').pop();
            let autoplayDismissed = false;
            if (queueVideoIds[queueVideoIds.length - 1] !== msgPayloadLastVideoId) {
                session.sendMessage(new Message.AutoplayUpNext(AID, null))
                    .catch((error) => __classPrivateFieldGet(this, _YouTubeApp_logger, "f").error('[yt-cast-receiver] Caught error sending message:', error));
                ;
                autoplayDismissed = true;
            }
            const stateBeforeSet = __classPrivateFieldGet(this, _YouTubeApp_player, "f").queue.getState();
            if (autoplayDismissed) {
                stateBeforeSet.autoplay = null;
            }
            const navBeforeSet = __classPrivateFieldGet(this, _YouTubeApp_player, "f").getNavInfo();
            await __classPrivateFieldGet(this, _YouTubeApp_player, "f").queue.updateByMessage(message, client);
            const stateAfterSet = __classPrivateFieldGet(this, _YouTubeApp_player, "f").queue.getState();
            const navAfterSet = __classPrivateFieldGet(this, _YouTubeApp_player, "f").getNavInfo();
            if (stateBeforeSet.autoplay?.id !== stateAfterSet.autoplay?.id) {
                sendMessages.push(new Message.AutoplayUpNext(AID, stateAfterSet.autoplay?.id || null));
            }
            if (message.name === 'setPlaylist' && (stateBeforeSet.current?.id !== stateAfterSet.current?.id ||
                stateBeforeSet.current?.context?.index !== stateAfterSet.current?.context?.index)) {
                if (__classPrivateFieldGet(this, _YouTubeApp_player, "f").status !== PLAYER_STATUSES.STOPPED) {
                    await __classPrivateFieldGet(this, _YouTubeApp_player, "f").stop(AID);
                }
                const currentVideo = stateAfterSet.current;
                if (currentVideo) {
                    await __classPrivateFieldGet(this, _YouTubeApp_player, "f").play(currentVideo, parseInt(payload.currentTime, 10) || 0, AID);
                }
            }
            else if (message.name === 'updatePlaylist' && !stateAfterSet.current) {
                await __classPrivateFieldGet(this, _YouTubeApp_player, "f").stop(AID);
            }
            else {
                sendMessages.push(new Message.NowPlaying(AID, await __classPrivateFieldGet(this, _YouTubeApp_player, "f").getState()));
                if (navBeforeSet.hasNext !== navAfterSet.hasNext || navBeforeSet.hasPrevious !== navAfterSet.hasPrevious) {
                    sendMessages.push(new Message.OnHasPreviousNextChanged(AID, navAfterSet));
                }
            }
            break;
        }
        case 'next': {
            if (!isSessionActive)
                return;
            await __classPrivateFieldGet(this, _YouTubeApp_player, "f").next(AID);
            break;
        }
        case 'previous': {
            if (!isSessionActive)
                return;
            await __classPrivateFieldGet(this, _YouTubeApp_player, "f").previous(AID);
            break;
        }
        case 'pause': {
            if (!isSessionActive)
                return;
            await __classPrivateFieldGet(this, _YouTubeApp_player, "f").pause(AID);
            break;
        }
        case 'stopVideo': {
            if (!isSessionActive)
                return;
            await __classPrivateFieldGet(this, _YouTubeApp_player, "f").stop(AID);
            break;
        }
        case 'seekTo': {
            if (!isSessionActive)
                return;
            await __classPrivateFieldGet(this, _YouTubeApp_player, "f").seek(parseInt(payload.newTime, 10));
            break;
        }
        case 'getVolume': {
            const volume = await __classPrivateFieldGet(this, _YouTubeApp_player, "f").getVolume();
            sendMessages.push(new Message.OnVolumeChanged(AID, volume));
            break;
        }
        case 'setVolume': {
            if (!isSessionActive)
                return;
            let mutedBool = false;
            if (typeof payload.muted === 'string') {
                mutedBool = payload.muted.toLowerCase().trim() === 'true';
            }
            else if (typeof payload.muted === 'boolean') {
                mutedBool = payload.muted;
            }
            const newVolume = {
                level: parseInt(payload.volume, 10),
                muted: mutedBool
            };
            const currentVolume = await __classPrivateFieldGet(this, _YouTubeApp_player, "f").getVolume();
            if (newVolume.level !== currentVolume.level || newVolume.muted !== currentVolume.muted) {
                await __classPrivateFieldGet(this, _YouTubeApp_player, "f").setVolume(newVolume, AID);
            }
            break;
        }
        case 'play': {
            if (!isSessionActive)
                return;
            await __classPrivateFieldGet(this, _YouTubeApp_player, "f").resume(AID);
            break;
        }
        case 'setAutoplayMode': {
            if (!isSessionActive)
                return;
            await __classPrivateFieldGet(this, _YouTubeApp_instances, "m", _YouTubeApp_setAutoplayMode).call(this, AID, payload.autoplayMode);
            break;
        }
        default:
            __classPrivateFieldGet(this, _YouTubeApp_logger, "f").debug(`[yt-cast-receiver] (AID: ${AID}) (${client.name}) Not handled: '${name}'`);
    }
    if (sendMessages.length > 0) {
        session.sendMessage(sendMessages)
            .catch((error) => __classPrivateFieldGet(this, _YouTubeApp_logger, "f").error('[yt-cast-receiver] Caught error sending message:', error));
        ;
    }
}, _YouTubeApp_handleSenderConnected = async function _YouTubeApp_handleSenderConnected(sender, session, AID) {
    await __classPrivateFieldGet(this, _YouTubeApp_instances, "m", _YouTubeApp_checkAndSwitchActiveSession).call(this, session);
    const sendMessages = [];
    const client = session.client;
    if (!__classPrivateFieldGet(this, _YouTubeApp_connectedSenders, "f").find((c) => c.id === sender.id)) {
        __classPrivateFieldGet(this, _YouTubeApp_logger, "f").info(`[yt-cast-receiver] (${client.name}) Sender connected: ${sender.name} (user: ${sender.user?.name})`);
        __classPrivateFieldGet(this, _YouTubeApp_logger, "f").debug(`[yt-cast-receiver] (${client.name}) Connected sender info:`, sender);
        __classPrivateFieldSet(this, _YouTubeApp_implicitlyDisconnectedSenders, __classPrivateFieldGet(this, _YouTubeApp_implicitlyDisconnectedSenders, "f").filter((s) => s.id !== sender.id), "f");
        __classPrivateFieldGet(this, _YouTubeApp_connectedSenders, "f").push(sender);
        await __classPrivateFieldGet(this, _YouTubeApp_instances, "m", _YouTubeApp_setAutoplayModeBySenderCapabilities).call(this, AID);
        await __classPrivateFieldGet(this, _YouTubeApp_instances, "m", _YouTubeApp_enforceMutePolicy).call(this, AID);
        const playerNavInfo = __classPrivateFieldGet(this, _YouTubeApp_player, "f").getNavInfo();
        const playerState = await __classPrivateFieldGet(this, _YouTubeApp_player, "f").getState();
        sendMessages.push(new Message.OnHasPreviousNextChanged(AID, playerNavInfo), new Message.NowPlaying(AID, playerState), new Message.OnStateChange(AID, playerState));
        this.emit('senderConnect', sender);
    }
    else {
        __classPrivateFieldGet(this, _YouTubeApp_logger, "f").debug(`[yt-cast-receiver] (${client.name}) Sender already connected.`);
    }
    return sendMessages;
}, _YouTubeApp_handleSenderDisconnected = async function _YouTubeApp_handleSenderDisconnected(sender, session, implicit, AID) {
    const isSessionActive = session === __classPrivateFieldGet(this, _YouTubeApp_activeSession, "f");
    if (!isSessionActive)
        return [];
    const client = session.client;
    const senderIndex = __classPrivateFieldGet(this, _YouTubeApp_connectedSenders, "f").findIndex((c) => c.id === sender.id);
    if (senderIndex < 0) {
        __classPrivateFieldGet(this, _YouTubeApp_logger, "f").warn(`[yt-cast-receiver] (${client.name}) Anomaly detected while unregistering disconnected sender: unable to find target among connected senders. Target:`, sender, ' Connected senders:', __classPrivateFieldGet(this, _YouTubeApp_connectedSenders, "f"));
        return [];
    }
    __classPrivateFieldGet(this, _YouTubeApp_connectedSenders, "f").splice(senderIndex, 1);
    __classPrivateFieldGet(this, _YouTubeApp_logger, "f").info(`[yt-cast-receiver] (${client.name}) Sender disconnected${implicit ? ' (implicit)' : ''}: ${sender.name}`);
    __classPrivateFieldGet(this, _YouTubeApp_logger, "f").debug(`[yt-cast-receiver] (${client.name}) Disconnected sender info:`, sender);
    if (implicit && !__classPrivateFieldGet(this, _YouTubeApp_implicitlyDisconnectedSenders, "f").some((s) => s.id === sender.id)) {
        __classPrivateFieldGet(this, _YouTubeApp_implicitlyDisconnectedSenders, "f").push(sender);
    }
    if (__classPrivateFieldGet(this, _YouTubeApp_connectedSenders, "f").length === 0) {
        __classPrivateFieldGet(this, _YouTubeApp_logger, "f").debug(`[yt-cast-receiver] No remaining senders connected. Reset player on disconnect policy is '${__classPrivateFieldGet(this, _YouTubeApp_resetPlayerOnDisconnectPolicy, "f")}'.`);
        let resetPlayer = false;
        if (__classPrivateFieldGet(this, _YouTubeApp_resetPlayerOnDisconnectPolicy, "f") === RESET_PLAYER_ON_DISCONNECT_POLICIES.ALL_DISCONNECTED) {
            resetPlayer = true;
        }
        else if (__classPrivateFieldGet(this, _YouTubeApp_resetPlayerOnDisconnectPolicy, "f") === RESET_PLAYER_ON_DISCONNECT_POLICIES.ALL_EXPLICITLY_DISCONNECTED) {
            if (__classPrivateFieldGet(this, _YouTubeApp_implicitlyDisconnectedSenders, "f").length > 0) {
                __classPrivateFieldGet(this, _YouTubeApp_logger, "f").debug(`[yt-cast-receiver] Not resetting player because there is one or more implicitly disconnected sender (count: ${__classPrivateFieldGet(this, _YouTubeApp_implicitlyDisconnectedSenders, "f").length}).`);
            }
            else {
                __classPrivateFieldGet(this, _YouTubeApp_logger, "f").debug('[yt-cast-receiver] No implicitly disconnected senders.');
                resetPlayer = true;
            }
        }
        if (resetPlayer) {
            __classPrivateFieldGet(this, _YouTubeApp_logger, "f").debug('[yt-cast-receiver] Resetting player...');
            await __classPrivateFieldGet(this, _YouTubeApp_player, "f").reset();
        }
    }
    else if (__classPrivateFieldGet(this, _YouTubeApp_player, "f").autoplayMode === AUTOPLAY_MODES.UNSUPPORTED) {
        await __classPrivateFieldGet(this, _YouTubeApp_instances, "m", _YouTubeApp_setAutoplayModeBySenderCapabilities).call(this, AID);
    }
    if (__classPrivateFieldGet(this, _YouTubeApp_connectedSenders, "f").length > 0) {
        await __classPrivateFieldGet(this, _YouTubeApp_instances, "m", _YouTubeApp_enforceMutePolicy).call(this, AID);
    }
    this.emit('senderDisconnect', sender, implicit);
    return [];
}, _YouTubeApp_handlePlayerStateEvent = function _YouTubeApp_handlePlayerStateEvent(payload) {
    if (!__classPrivateFieldGet(this, _YouTubeApp_activeSession, "f")) {
        return;
    }
    const { AID, current, previous } = payload;
    if (__classPrivateFieldGet(this, _YouTubeApp_connectedSenders, "f").length === 0) {
        __classPrivateFieldGet(this, _YouTubeApp_logger, "f").debug('[yt-cast-receiver] Ignoring player state event because there is no connected sender.');
        return;
    }
    __classPrivateFieldGet(this, _YouTubeApp_logger, "f").debug('[yt-cast-receiver] Player state changed from:', previous);
    __classPrivateFieldGet(this, _YouTubeApp_logger, "f").debug('To:', current);
    let statusChanged = true, positionChanged = true, volumeChanged = true, nowPlayingChanged = true, autoplayChanged = true;
    if (previous) {
        statusChanged = previous.status !== current.status;
        positionChanged = previous.position !== current.position;
        volumeChanged = previous.volume.level !== current.volume.level || previous.volume.muted !== current.volume.muted;
        nowPlayingChanged = previous.queue.current?.id !== current.queue.current?.id ||
            previous.queue.id !== current.queue.id ||
            previous.queue.current?.context?.index !== current.queue.current?.context?.index;
        autoplayChanged = previous.queue.autoplay?.id !== current.queue.autoplay?.id;
    }
    const messages = [];
    if (statusChanged || positionChanged) {
        messages.push(new Message.OnStateChange(AID, current));
    }
    if (nowPlayingChanged || (statusChanged && (!previous || previous.status !== PLAYER_STATUSES.PLAYING))) {
        messages.push(new Message.NowPlaying(AID, current), new Message.OnHasPreviousNextChanged(AID, __classPrivateFieldGet(this, _YouTubeApp_player, "f").getNavInfo()));
    }
    if (volumeChanged) {
        messages.push(new Message.OnVolumeChanged(AID, current.volume));
    }
    if (autoplayChanged) {
        messages.push(new Message.AutoplayUpNext(AID, current.queue.autoplay?.id || null));
    }
    if (messages.length > 0) {
        if (messages.every((c) => c instanceof Message.OnVolumeChanged && c.AID !== null)) {
            // We could be responding to a series of 'setVolume' messages. To reduce
            // Lag on sender side, we only send the latest 'onVolumeChanged' message after
            // A short interval of no further such messages.
            __classPrivateFieldGet(this, _YouTubeApp_activeSession, "f").sendMessage(messages, { key: 'onVolumeChanged', interval: 200 })
                .catch((error) => __classPrivateFieldGet(this, _YouTubeApp_logger, "f").error('[yt-cast-receiver] Caught error sending message:', error));
            ;
        }
        else {
            __classPrivateFieldGet(this, _YouTubeApp_activeSession, "f").sendMessage(messages)
                .catch((error) => __classPrivateFieldGet(this, _YouTubeApp_logger, "f").error('[yt-cast-receiver] Caught error sending message:', error));
            ;
        }
    }
};
export default YouTubeApp;
