// yt-cast-receiver/dist/lib/dial/DialServer.js
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _DialServer_dialOptions, _DialServer_expressServer, _DialServer_dialServer, _DialServer_logger, _DialServer_status;
import express from 'express';
import http from 'http';
import dial from '@patrickkfkan/peer-dial';
import { DialServerError } from '../utils/Errors.js';
import { CONF_DEFAULTS, STATUSES } from '../Constants.js';
function createDelegate(apps, logger) {
    return {
        getApp: (appName) => {
            return apps[appName];
        },
        /**
         * Called when a sender connects through DIAL. At this point, YouTubeApp is already running.
         * `launchData` contains the pairing code, which YouTubeApp uses for pairing with the sender.
         * @param appName - `YouTube` - anything else will throw an error.
         * @param launchData - Contains the pairing code which YouTubeApp uses to pair with sender.
         * @param callback - Function to provide `pid` back to the DIAL server.
         */
        launchApp: (appName, launchData, callback) => {
            if (apps[appName]) {
                apps[appName].launch(launchData).then((pid) => {
                    callback(pid);
                })
                    .catch((error) => {
                    logger.error('[yt-cast-receiver] DIAL server failed to launch app:', error);
                    callback('');
                });
            }
            else {
                throw new DialServerError(`Unknown app name: ${appName}`);
            }
        },
        /**
         * This should never get called because `YouTubeApp.allowStop` is `false`. We stop YouTubeApp
         * manually in `YouTubeCastReceiver.stop()`.
         */
        stopApp: (_appName, _pid, _callback) => {
            // Do nothing
        }
    };
}
/**
 * Broadcasts receiver device on the network so that YouTube sender apps can initiate
 * Cast sessions through the DIAL protocol.
 */
class DialServer {
    constructor(app, options) {
        _DialServer_dialOptions.set(this, void 0);
        _DialServer_expressServer.set(this, void 0);
        _DialServer_dialServer.set(this, void 0);
        _DialServer_logger.set(this, void 0);
        _DialServer_status.set(this, void 0);
        __classPrivateFieldSet(this, _DialServer_logger, options.logger, "f");
        const expressApp = express();
        __classPrivateFieldSet(this, _DialServer_dialOptions, {
            expressApp,
            port: options.port || 3000,
            corsAllowOrigins: options.corsAllowOrigins || false,
            prefix: options.prefix || '/ytcr',
            friendlyName: options.friendlyName,
            manufacturer: options.manufacturer || CONF_DEFAULTS.BRAND,
            modelName: options.modelName || CONF_DEFAULTS.MODEL,
            delegate: createDelegate({ 'YouTube': app }, __classPrivateFieldGet(this, _DialServer_logger, "f")),
            bindToInterfaces: options.bindToInterfaces,
            bindToAddresses: options.bindToAddresses
        }, "f");
        __classPrivateFieldSet(this, _DialServer_dialServer, new dial.Server(__classPrivateFieldGet(this, _DialServer_dialOptions, "f")), "f");
        __classPrivateFieldSet(this, _DialServer_expressServer, http.createServer(expressApp), "f");
        __classPrivateFieldSet(this, _DialServer_status, STATUSES.STOPPED, "f");
    }
    start() {
        if (__classPrivateFieldGet(this, _DialServer_status, "f") !== STATUSES.STOPPED) {
            __classPrivateFieldGet(this, _DialServer_logger, "f").warn('[yt-cast-receiver] start() called but DIAL server not in STOPPED state.');
            return Promise.resolve();
        }
        __classPrivateFieldSet(this, _DialServer_status, STATUSES.STARTING, "f");
        __classPrivateFieldGet(this, _DialServer_logger, "f").debug('[yt-cast-receiver] Starting DIAL server...');
        return new Promise((resolve, reject) => {
            try {
                __classPrivateFieldGet(this, _DialServer_expressServer, "f").listen(__classPrivateFieldGet(this, _DialServer_dialOptions, "f").port, () => {
                    __classPrivateFieldGet(this, _DialServer_dialServer, "f").start();
                    __classPrivateFieldSet(this, _DialServer_status, STATUSES.RUNNING, "f");
                    __classPrivateFieldGet(this, _DialServer_logger, "f").info(`[yt-cast-receiver] DIAL server listening on port ${__classPrivateFieldGet(this, _DialServer_dialOptions, "f").port}`);
                    resolve();
                }).on('error', (error) => {
                    __classPrivateFieldSet(this, _DialServer_status, STATUSES.STOPPED, "f");
                    reject(new DialServerError('Failed to start DIAL server', error));
                });
            }
            catch (error) {
                __classPrivateFieldSet(this, _DialServer_status, STATUSES.STOPPED, "f");
                reject(new DialServerError('Failed to start DIAL server', error));
            }
        });
    }
    stop() {
        if (__classPrivateFieldGet(this, _DialServer_status, "f") !== STATUSES.RUNNING) {
            __classPrivateFieldGet(this, _DialServer_logger, "f").warn('[yt-cast-receiver] stop() called but DIAL server not in RUNNING state.');
            return Promise.resolve();
        }
        __classPrivateFieldSet(this, _DialServer_status, STATUSES.STOPPING, "f");
        __classPrivateFieldGet(this, _DialServer_logger, "f").debug('[yt-cast-receiver] Stopping DIAL server...');
        return new Promise((resolve, reject) => {
            try {
                __classPrivateFieldGet(this, _DialServer_dialServer, "f").stop();
                __classPrivateFieldGet(this, _DialServer_expressServer, "f").close();
                __classPrivateFieldSet(this, _DialServer_status, STATUSES.STOPPED, "f");
                resolve();
            }
            catch (error) {
                __classPrivateFieldSet(this, _DialServer_status, STATUSES.RUNNING, "f");
                reject(new DialServerError('Failed to stop DIAL server', error));
            }
        });
    }
    get status() {
        return __classPrivateFieldGet(this, _DialServer_status, "f");
    }
}
_DialServer_dialOptions = new WeakMap(), _DialServer_expressServer = new WeakMap(), _DialServer_dialServer = new WeakMap(), _DialServer_logger = new WeakMap(), _DialServer_status = new WeakMap();
export default DialServer;
